{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1ef8b2d1",
   "metadata": {},
   "source": [
    "# Import and define functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ceba86f0",
   "metadata": {},
   "source": [
    "Installs packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "596f1777",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"!jupyter nbextension enable --py --sys-prefix qgrid\n",
    "!jupyter nbextension enable --py widgetsnbextension\n",
    "!pip install qgrid --upgrade\n",
    "!pip install notebook --upgrade\n",
    "!pip install nodejs\n",
    "!pip install npm\n",
    "!jupyter labextension install qgrid2\n",
    "!pip install --upgrade openai\n",
    "!pip install python-dotenv\"\"\";"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e649b40a",
   "metadata": {},
   "source": [
    "Imports libraries and OpenAI key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9e121e59",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "# Set your email address for PubMed API (replace with your email)\n",
    "Entrez.email = \"kuhfeldrf@oregonstate.edu\"\n",
    "import requests\n",
    "import xml.etree.ElementTree as ET\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import time\n",
    "import os\n",
    "import string\n",
    "import qgrid\n",
    "import os\n",
    "#from dotenv import load_dotenv\n",
    "import openai\n",
    "#os.environ['OPENAI_API_KEY'] = \"sk-..SI\"\n",
    "#os.environ.get('OPENAI_API_KEY')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb3b017e",
   "metadata": {},
   "source": [
    "<b>Uniport Function: </b>Extracts the protein sequence information from the Uniprot protein page"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a519a5bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_sequence(lines):\n",
    "    \"\"\"\n",
    "    Extracts the protein sequence information from the Uniprot protein page\n",
    "    \"\"\"\n",
    "    protein_sequence, chain,protein_desc, protein_name = '', '','',''\n",
    "    protein_desc_list = []\n",
    "    protein_name_list = []\n",
    "    # Regular expressions\n",
    "    sequence_pattern = re.compile(r'<sequence[^>]*>(.*?)</sequence>')\n",
    "    #chain_pattern = re.compile(r'<feature type=\"chain\".*description=\"(.*?)\"')\n",
    "    protein_desc_pattern = re.compile(r'<fullName.*?>(.*?)</fullName>')\n",
    "    protein_name_pattern = re.compile(r'<name>(.*?)</name>')\n",
    "\n",
    "    for line in lines:\n",
    "        if '<name' in line:\n",
    "            match = protein_name_pattern.search(line)\n",
    "            if match:\n",
    "                protein_name_list.append(match.group(1))\n",
    "        if not protein_desc and '<fullName' in line:\n",
    "            match = protein_desc_pattern.search(line)\n",
    "            if match:\n",
    "                protein_desc_list.append(match.group(1))\n",
    "        if '<sequence' in line:\n",
    "            match = sequence_pattern.search(line)\n",
    "            if match:\n",
    "                protein_sequence = match.group(1)\n",
    "        #if '<feature type=\"chain\"' in line:\n",
    "        #    match = chain_pattern.search(line)\n",
    "        #    if match:\n",
    "        #        chain = match.group(1)\n",
    "    if protein_name_list:     \n",
    "        protein_name=protein_name_list[0]\n",
    "    if protein_desc_list:     \n",
    "        protein_desc=protein_desc_list[0]\n",
    "\n",
    "    \n",
    "    #return protein_sequence, chain, protein_name, protein_desc\n",
    "    return protein_sequence, protein_name, protein_desc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9fcfef1",
   "metadata": {},
   "source": [
    "<b>Uniport Function: </b>Extract a mapping of evidence keys to their associated PubMed IDs from the XML root node."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5e91d13b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_evidence_to_pubmed_mapping(root):\n",
    "    \"\"\"\n",
    "    Extract a mapping of evidence keys to their associated PubMed IDs from the XML root node.\n",
    "\n",
    "    Args:\n",
    "    - root (xml.etree.ElementTree.Element): The root node of the XML document.\n",
    "\n",
    "    Returns:\n",
    "    - dict: A dictionary mapping evidence keys to PubMed IDs.\n",
    "    \"\"\"\n",
    "    evidence_to_pubmed = {}\n",
    "    \n",
    "    # Extract evidence mapping to PubMed ID\n",
    "    for evidence in root.findall(\".//{http://uniprot.org/uniprot}evidence\"):\n",
    "        evidence_key = evidence.attrib.get('key')\n",
    "        source = evidence.find(\".//{http://uniprot.org/uniprot}dbReference[@type='PubMed']\")\n",
    "        \n",
    "        if source is not None:\n",
    "            pubmed_id = source.attrib.get('id')\n",
    "            if evidence_key and pubmed_id:\n",
    "                evidence_to_pubmed[evidence_key] = pubmed_id\n",
    "    \n",
    "    return evidence_to_pubmed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8174b4c4",
   "metadata": {},
   "source": [
    "<b>Uniport Function: </b>Extracts peptide features and function comments from a given XML data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "894bcb72",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_peptide_and_function(data, protein_id):\n",
    "    \"\"\"\n",
    "    Extracts peptide features and function comments from a given XML data.\n",
    "    \"\"\"\n",
    "    # Parsing XML content\n",
    "    root = ET.fromstring(data)\n",
    "    lines = data.split('\\n')\n",
    "    #protein_sequence, chain, protein_name, protein_desc = fetch_sequence(lines)\n",
    "    protein_sequence, protein_name, protein_desc = fetch_sequence(lines)\n",
    "    # Extract the mapping from evidence key to PubMed ID\n",
    "    evidence_to_pubmed = extract_evidence_to_pubmed_mapping(root)\n",
    "    # Extract peptide features\n",
    "    peptide_features = []\n",
    "    for feature in root.findall(\".//{http://uniprot.org/uniprot}feature[@type='peptide']\"):\n",
    "        begin = int(feature.find(\"{http://uniprot.org/uniprot}location/{http://uniprot.org/uniprot}begin\").attrib.get('position', ''))\n",
    "        end = int(feature.find(\"{http://uniprot.org/uniprot}location/{http://uniprot.org/uniprot}end\").attrib.get('position', ''))\n",
    "        peptide_seq = protein_sequence[begin-1:end]  # Extract peptide sequence using begin and end\n",
    "        interval = f'{begin}-{end}'\n",
    "        peptide_info = {\n",
    "            'proteinID': protein_id,\n",
    "            'protein_name': protein_name,\n",
    "            'protein_desc':protein_desc,\n",
    "            #'chain': chain,\n",
    "            'interval':interval,\n",
    "            'peptide': peptide_seq,\n",
    "            'description': feature.attrib.get('description', None),\n",
    "            'feature_evidence': feature.attrib.get('evidence', None),\n",
    "            'evidence_to_pubmed': evidence_to_pubmed,\n",
    "            #'begin': begin,\n",
    "            #'end': end,\n",
    "            #'id': feature.attrib.get('id', None),\n",
    "            #'protein_sequence': protein_sequence,\n",
    "        }\n",
    "        peptide_features.append(peptide_info)\n",
    "    \n",
    "    # Extract function comments\n",
    "    function_comments = []\n",
    "    for comment in root.findall(\".//{http://uniprot.org/uniprot}comment[@type='function']\"):\n",
    "        function_info = {\n",
    "            'molecule': comment.find(\"{http://uniprot.org/uniprot}molecule\").text if comment.find(\"{http://uniprot.org/uniprot}molecule\") is not None else None,\n",
    "            'text': comment.find(\"{http://uniprot.org/uniprot}text\").text,\n",
    "            'comment_evidence': comment.find(\"{http://uniprot.org/uniprot}text\").attrib.get('evidence', None)\n",
    "        }\n",
    "        function_comments.append(function_info)\n",
    "    \n",
    "    return peptide_features, function_comments, evidence_to_pubmed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "372b26ab",
   "metadata": {},
   "source": [
    "<b>Uniport Function: </b>To associate peptide features with related function comments based on shared evidence or matching descriptions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e8ca82a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def associate_peptide_with_function(peptide_features, function_comments, evidence_to_pubmed):\n",
    "    \"\"\"\n",
    "    To associate peptide features with related function comments based on shared evidence or matching descriptions.\n",
    "    \"\"\"\n",
    "    associated_data = []\n",
    "    \n",
    "    for peptide in peptide_features:\n",
    "        associated_comment_texts = []\n",
    "        all_functions = []  # List to store all functions if not directly linked\n",
    "        peptide_evidence = set(peptide['feature_evidence'].split()) if peptide['feature_evidence'] else set()\n",
    "        \n",
    "        # Fetch associated PubMed IDs based on evidence keys that are also present in the evidence column\n",
    "        peptide_pubmed = {evidence_to_pubmed[eid] for eid in peptide_evidence if eid in evidence_to_pubmed}\n",
    "        \n",
    "        direct_link = False  # Variable to track if there's a direct link for the peptide\n",
    "        \n",
    "        for comment in function_comments:\n",
    "            comment_evidence = set(comment['comment_evidence'].split()) if comment['comment_evidence'] else set()\n",
    "            # Check for shared evidence or if peptide description matches comment molecule\n",
    "            if peptide_evidence & comment_evidence or (peptide['description'] == comment.get('molecule')):\n",
    "                peptide['comment_evidence'] = comment_evidence\n",
    "                associated_comment_texts.append(comment['text'])\n",
    "                direct_link = True\n",
    "                # Add PubMed IDs associated with the comment's evidence to peptide_pubmed\n",
    "                comment_pubmed = {evidence_to_pubmed[eid] for eid in comment_evidence if eid in evidence_to_pubmed}\n",
    "                peptide_pubmed.update(comment_pubmed)\n",
    "            else:\n",
    "                all_functions.append(comment['text'])\n",
    "        \n",
    "        peptide['associated_function'] = '; '.join(associated_comment_texts)\n",
    "        if not direct_link and all_functions:\n",
    "            peptide['non_associated_function'] = all_functions\n",
    "        \n",
    "        # Convert the set to a list\n",
    "        peptide['evidence_pubmed'] = list(peptide_pubmed)\n",
    "        associated_data.append(peptide)\n",
    "    \n",
    "    return associated_data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ee526bd",
   "metadata": {},
   "source": [
    "<b>Uniport Function: </b>To extract references from the provided XML data and return them as a DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "29f63c35",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_references(data, protein_id):\n",
    "    \"\"\"\n",
    "    To extract references from the provided XML data and return them as a DataFrame.\n",
    "    \"\"\"\n",
    "    lines = data.split('\\n')\n",
    "\n",
    "    # Continue extracting reference data\n",
    "    reference_data = []\n",
    "    inside_reference = False\n",
    "    title, pubmed_id, doi, reference_key, formatted_authors, scope_range = None, None, None, None, None, None\n",
    "    author_list = []\n",
    "\n",
    "    # Regular expressions for extracting information\n",
    "    reference_key_pattern = re.compile(r'<reference key=\"(\\d+)\">')\n",
    "    title_pattern = re.compile(r'<title>(.*?)</title>')\n",
    "    person_name_pattern = re.compile(r'<person name=\"(.*?)\"/>')\n",
    "    dbReference_type_pattern = re.compile(r'<dbReference type=\"(.*?)\"')\n",
    "    dbReference_id_pattern = re.compile(r'id=\"(.*?)\"/>')\n",
    "    scope_pattern = re.compile(r'PROTEIN SEQUENCE OF (\\d+-\\d+)')\n",
    "\n",
    "    for line in lines:\n",
    "        try:\n",
    "            # Extract reference information\n",
    "            if '<reference key=\"' in line:\n",
    "                inside_reference = True\n",
    "                reference_key = reference_key_pattern.search(line).group(1)\n",
    "                # Ensure author_list is always initialized as an empty list\n",
    "                author_list = []\n",
    "            elif inside_reference:\n",
    "                if '<title>' in line:\n",
    "                    title = title_pattern.search(line).group(1)\n",
    "                elif '<person name=\"' in line:\n",
    "                    author = person_name_pattern.search(line).group(1)\n",
    "                    author_list.append(author)\n",
    "                elif '<dbReference type=\"' in line:\n",
    "                    ref_type = dbReference_type_pattern.search(line).group(1)\n",
    "                    ref_id = dbReference_id_pattern.search(line).group(1)\n",
    "                    if ref_type == \"PubMed\":\n",
    "                        pubmed_id = ref_id\n",
    "                    elif ref_type == \"DOI\":\n",
    "                        doi = ref_id\n",
    "                elif '<scope>' in line:\n",
    "                    scope_match = scope_pattern.search(line)\n",
    "                    if scope_match:\n",
    "                        scope_range = scope_match.group(1)\n",
    "                elif '</reference>' in line:\n",
    "                    inside_reference = False\n",
    "                    # Format the authors list\n",
    "                    if author_list:\n",
    "                        if len(author_list) >= 1:\n",
    "                            last_name, first_name_initial = author_list[0].split(' ')[0], author_list[0].split(' ')[1][0]\n",
    "                            formatted_authors = f\"{last_name}, {first_name_initial}. et al.\"\n",
    "                    else:\n",
    "                        formatted_authors = None\n",
    "\n",
    "                    reference_data.append({\n",
    "                        'proteinID': protein_id,\n",
    "                        'reference key': reference_key,\n",
    "                        'title': title,\n",
    "                        'authors': formatted_authors,\n",
    "                        'pubmed': pubmed_id,\n",
    "                        'doi': doi,\n",
    "                        'scope_range': scope_range\n",
    "                    })\n",
    "\n",
    "                    # Reset for the next reference\n",
    "                    author_list, pubmed_id, doi, title, reference_key, formatted_authors, scope_range = [], None, None, None, None, None, None \n",
    "        except Exception as e:\n",
    "            print(f\"Error processing line: {line}. Error: {e}\")\n",
    "\n",
    "    # Convert reference_data list to a DataFrame\n",
    "    \n",
    "    return pd.DataFrame(reference_data)  # Note that this should return reference_df, not reference_data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73393ba0",
   "metadata": {},
   "source": [
    "<b>Uniport Function: </b>Fetch protein information from UniProt for a given protein ID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3124b7f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_protein_info(protein_id):\n",
    "    \"\"\"\n",
    "    Fetch protein information from UniProt for a given protein ID.\n",
    "    \"\"\"\n",
    "    url = f'https://www.uniprot.org/uniprot/{protein_id}.xml'\n",
    "    response = requests.get(url)\n",
    "\n",
    "    # Check if the request was successful\n",
    "    if response.status_code != 200:\n",
    "        return pd.DataFrame(), pd.DataFrame()\n",
    "    \n",
    "    data = response.text\n",
    "\n",
    "    # Extract peptide features and function comments\n",
    "    peptide_features, function_comments, evidence_to_pubmed = extract_peptide_and_function(data, protein_id)\n",
    "    # Extract reference data\n",
    "    reference_df = extract_references(data, protein_id)\n",
    "\n",
    "    # Associate peptide with function\n",
    "    associated_data = associate_peptide_with_function(peptide_features, function_comments, evidence_to_pubmed)\n",
    "\n",
    "    # Convert associated_data list to a DataFrame\n",
    "    peptide_df = pd.DataFrame(associated_data)\n",
    "\n",
    "    return peptide_df, reference_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d10096e3",
   "metadata": {},
   "source": [
    "<b>Uniport Function: </b>Update associated function updates the asscoaitated function column from the non_Asscoiated_function matching pubmed ID or description column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0fa4abce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_associated_function(row):\n",
    "    # If associated_function is NaN or an empty string, proceed with the update\n",
    "    if pd.isna(row['associated_function']) or row['associated_function'] == '':\n",
    "        if isinstance(row['non_associated_function'], str) and row['non_associated_function'].startswith('[') and row['non_associated_function'].endswith(']'):\n",
    "            non_associated_entries = eval(row['non_associated_function'])\n",
    "        else:\n",
    "            non_associated_entries = row['non_associated_function'] if isinstance(row['non_associated_function'], list) else []\n",
    "\n",
    "        description = row['description'].lower().replace(\"-\", \" \") if pd.notna(row['description']) else \"\"\n",
    "        associated_info = \"\"\n",
    "        \n",
    "        # Check if the DataFrame has a 'pubmed_id' column\n",
    "        pubmed_id = str(row['pubmed_id']) if 'pubmed_id' in row.keys() and pd.notna(row['pubmed_id']) else \"-1\" #-1 represents a value that a pubmedID will never be\n",
    "\n",
    "        for entry in non_associated_entries:\n",
    "            # Check if entry contains the description or the pubmed_id\n",
    "            if description in entry.lower().replace(\"-\", \" \") or pubmed_id in entry:\n",
    "                associated_info = associated_info + ' ' + entry if associated_info else entry\n",
    "                \n",
    "        return associated_info\n",
    "    else:\n",
    "        return row['associated_function']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab5803a",
   "metadata": {},
   "source": [
    "<b>PubMed Function: </b>Function to fetch abstract, title, authors, and doi for a PubMed ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "79ee8e26",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_pubmed_by_title(title):\n",
    "    handle = Entrez.esearch(db=\"pubmed\", term=title)\n",
    "    record = Entrez.read(handle)\n",
    "    return record[\"IdList\"]\n",
    "\n",
    "def fetch_details(row, row_num, total_rows):\n",
    "    \"\"\"\n",
    "    Function to fetch abstract, title, authors, and doi based on either PubMed ID, DOI, or title\n",
    "    \"\"\"\n",
    "    details = {}\n",
    "    pubmed_id = row['pubmed_id']\n",
    "    \n",
    "    # Check if pubmed_id is NaN or None\n",
    "    if pd.isnull(pubmed_id):\n",
    "        # Try to search by title\n",
    "        matching_ids = search_pubmed_by_title(row['title'])\n",
    "        if matching_ids:\n",
    "            pubmed_id = matching_ids[0]  # take the first matching ID\n",
    "        else:\n",
    "            print(f\"Row {row_num}/{total_rows}: No PubMed ID found for title '{row['title']}'\")\n",
    "            return None\n",
    "    \n",
    "    # Fetch details by PubMed ID\n",
    "    try:\n",
    "        handle = Entrez.efetch(db=\"pubmed\", id=pubmed_id, retmode=\"xml\")\n",
    "        record = Entrez.read(handle)\n",
    "        \n",
    "        # Fetch abstract if available\n",
    "        try:\n",
    "            details['abstract'] = record['PubmedArticle'][0]['MedlineCitation']['Article']['Abstract']['AbstractText'][0]\n",
    "        except:\n",
    "            details['abstract'] = None\n",
    "        \n",
    "        # Fetch title if available\n",
    "        try:\n",
    "            details['title'] = record['PubmedArticle'][0]['MedlineCitation']['Article']['ArticleTitle']\n",
    "        except:\n",
    "            details['title'] = None\n",
    "        \n",
    "        # Fetch authors if available\n",
    "        try:\n",
    "            author_list = [author['LastName'] + \", \" + author['Initials'] + \".\" for author in record['PubmedArticle'][0]['MedlineCitation']['Article']['AuthorList']]\n",
    "            # Format the authors\n",
    "            if len(author_list) > 1:\n",
    "                details['authors'] = author_list[0] + \" et al.\"\n",
    "            else:\n",
    "                details['authors'] = author_list[0]\n",
    "        except:\n",
    "            details['authors'] = None\n",
    "        \n",
    "        # Fetch doi if available\n",
    "        try:\n",
    "            for article_id in record['PubmedArticle'][0]['PubmedData']['ArticleIdList']:\n",
    "                if article_id.attributes['IdType'] == \"doi\":\n",
    "                    details['doi'] = article_id.title()\n",
    "                    break\n",
    "        except:\n",
    "            details['doi'] = None\n",
    "        \n",
    "        details['pubmed_id'] = pubmed_id        \n",
    "        print(f\"Row {row_num}/{total_rows}: Successfully fetched details for PubMed ID {pubmed_id}\")\n",
    "        return details\n",
    "    except Exception as e:\n",
    "        print(f\"Row {row_num}/{total_rows}: Error fetching details for PubMed ID {pubmed_id}: {str(e)}\")\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02afd60b",
   "metadata": {},
   "source": [
    "<b>Function: </b>Loops through a speciese specific list of proteins, fetching info from Uniprot, then returns two data frames with and without fetchable reference data named after a the provided species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0f2bac67",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_species_data(protein_ids_list):\n",
    "    \"\"\"\n",
    "    Refined version of process_species_data to avoid the SettingWithCopyWarning.\n",
    "    \"\"\"\n",
    "    reference_data_dfs = []\n",
    "    \n",
    "    peptide_data_dfs = []\n",
    "\n",
    "    # Iterate through the protein IDs and call the function\n",
    "    for index, protein_id in enumerate(protein_ids_list):\n",
    "        start_time = time.time()\n",
    "        peptide_data_df, reference_data_df = fetch_protein_info(protein_id)\n",
    "        elapsed_time = time.time() - start_time\n",
    "        peptide_data_dfs.append(peptide_data_df)\n",
    "        reference_data_dfs.append(reference_data_df)\n",
    "        print(f\"Processing {protein_id} (Index: {index + 1} of {len(protein_ids_list)}) took {elapsed_time:.2f} seconds.\")\n",
    "    \n",
    "    processed_peptide_df = pd.concat(peptide_data_dfs, ignore_index=True)\n",
    "    processed_reference_df = pd.concat(reference_data_dfs, ignore_index=True)\n",
    "    return processed_peptide_df, processed_reference_df   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f278d01e",
   "metadata": {},
   "source": [
    "<b>Function: </b>Prints inportant info on the dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ac7cd1d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_critical_info(df, df_name):\n",
    "    \"\"\"Print critical information about a DataFrame.\"\"\"\n",
    "    \n",
    "    print(f\"Information for DataFrame: {df_name}\")\n",
    "    print(\"-\" * 40)\n",
    "    \n",
    "    # Shape of the DataFrame\n",
    "    print(f\"Numbers of peptides: {df.shape[0]}\")\n",
    "    print(f\"Shape: {df.shape}\")\n",
    "    \n",
    "    # List of columns\n",
    "    print(f\"Columns: {df.columns.tolist()}\")\n",
    "    \n",
    "    # Data types of each column\n",
    "    #print(\"\\nData Types:\")\n",
    "    #print(df.dtypes)\n",
    "    \n",
    "    # Number of missing values in each column\n",
    "    print(\"\\nMissing Values Count:\")\n",
    "    print(df.isnull().sum())\n",
    "    \n",
    "    # Display the first few rows of the DataFrame\n",
    "    #print(\"\\nHead of DataFrame:\")\n",
    "    #print(df.head())\n",
    "    \n",
    "    print(\"\\n\" + \"=\" * 40 + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0cd34b6",
   "metadata": {},
   "source": [
    "# Bring in protein & MBPDB lists"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d27a1267",
   "metadata": {},
   "source": [
    "MBPDB list imported from Summer 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f753d550",
   "metadata": {},
   "outputs": [],
   "source": [
    "mbpdb_df = pd.read_csv('data\\exported_data.tsv', sep='\\t')\n",
    "mbpdb_df.rename(columns={'protein_pid': 'proteinID', 'intervals': 'interval'}, inplace=True)\n",
    "mbpdb_function_list = list(set(mbpdb_df['function']))\n",
    "mbpdb_df_qgrid = qgrid.show_grid(mbpdb_df, show_toolbar=True, grid_options={'forceFitColumns': False})\n",
    "#mbpdb_df_qgrid\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75abebfe",
   "metadata": {},
   "source": [
    "list of human milk proteins from OSU Proteome Discoverer Dave made from a few papers not comprehensive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ae2b1b29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "382\n"
     ]
    }
   ],
   "source": [
    "# Read the file\n",
    "with open('protein_lists\\HumanMilkProteinDatabase_v2.fasta', 'r') as file:\n",
    "    content = file.readlines()\n",
    "\n",
    "# Extract unique protein IDs\n",
    "hum_protein_ids = {line.split('|')[1] for line in content if line.startswith('>sp|')}\n",
    "hum_protein_ids=list(hum_protein_ids)\n",
    "print(len(hum_protein_ids))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7003503",
   "metadata": {},
   "source": [
    "Test lists and list of proteins found in MBPDB peptides used for developement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "103c163a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#list of 42 proteins that currently have matches in the MBPDB\n",
    "uniprot_ids = [    \"P02666\",    \"P47710\",    \"P02662\",    \"P04653\",    \"P09115\",    \"P18626\",    \"O97943\",    \"O62823\",    \"P02663\",    \"P04654\",   \"P33049\",    \"A0A1L6KYI1\",    \"E9NZN2\",    \"P05814\",    \"P11839\",    \"P09116\",    \"Q9TSI0\",    \"Q9TVD0\",    \"P33048\",    \"P86273\",    \"A0A344X7B9\",    \"P02668\",    \"P07498\",    \"P02669\",    \"I6UFY2\",    \"P02670\",    \"P80195\",    \"P02754\",    \"P02755\",    \"P02756\",    \"P00711\",    \"P00710\",    \"P24627\",    \"P02788\",    \"P14632\",    \"Q29477\",    \"O77698\",    \"P14639\",    \"P02769\",    \"P67976\",    \"L8I8G5\",    \"P01966\"]\n",
    "\n",
    "#short lits of proteins to test\n",
    "test_list=['P61278','O43612', 'P02788','P62158','P02666']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4807be8",
   "metadata": {},
   "source": [
    "# Executes code and searches list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4bff521",
   "metadata": {},
   "source": [
    "This is the code that runs the Uniport functions above given a protein list of interest, averages ~2 sec per protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "f31103a4",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing P02666 (Index: 1 of 42) took 1.70 seconds.\n",
      "Processing P47710 (Index: 2 of 42) took 1.62 seconds.\n",
      "Processing P02662 (Index: 3 of 42) took 1.62 seconds.\n",
      "Processing P04653 (Index: 4 of 42) took 1.62 seconds.\n",
      "Processing P09115 (Index: 5 of 42) took 1.64 seconds.\n",
      "Processing P18626 (Index: 6 of 42) took 1.58 seconds.\n",
      "Processing O97943 (Index: 7 of 42) took 1.57 seconds.\n",
      "Processing O62823 (Index: 8 of 42) took 1.59 seconds.\n",
      "Processing P02663 (Index: 9 of 42) took 1.58 seconds.\n",
      "Processing P04654 (Index: 10 of 42) took 1.54 seconds.\n",
      "Processing P33049 (Index: 11 of 42) took 1.65 seconds.\n",
      "Processing A0A1L6KYI1 (Index: 12 of 42) took 1.59 seconds.\n",
      "Processing E9NZN2 (Index: 13 of 42) took 1.59 seconds.\n",
      "Processing P05814 (Index: 14 of 42) took 1.57 seconds.\n",
      "Processing P11839 (Index: 15 of 42) took 1.62 seconds.\n",
      "Processing P09116 (Index: 16 of 42) took 1.60 seconds.\n",
      "Processing Q9TSI0 (Index: 17 of 42) took 1.57 seconds.\n",
      "Processing Q9TVD0 (Index: 18 of 42) took 1.65 seconds.\n",
      "Processing P33048 (Index: 19 of 42) took 1.54 seconds.\n",
      "Processing P86273 (Index: 20 of 42) took 1.56 seconds.\n",
      "Processing A0A344X7B9 (Index: 21 of 42) took 1.63 seconds.\n",
      "Processing P02668 (Index: 22 of 42) took 1.56 seconds.\n",
      "Processing P07498 (Index: 23 of 42) took 1.63 seconds.\n",
      "Processing P02669 (Index: 24 of 42) took 1.59 seconds.\n",
      "Processing I6UFY2 (Index: 25 of 42) took 1.60 seconds.\n",
      "Processing P02670 (Index: 26 of 42) took 1.62 seconds.\n",
      "Processing P80195 (Index: 27 of 42) took 1.58 seconds.\n",
      "Processing P02754 (Index: 28 of 42) took 1.60 seconds.\n",
      "Processing P02755 (Index: 29 of 42) took 1.59 seconds.\n",
      "Processing P02756 (Index: 30 of 42) took 1.60 seconds.\n",
      "Processing P00711 (Index: 31 of 42) took 1.61 seconds.\n",
      "Processing P00710 (Index: 32 of 42) took 1.55 seconds.\n",
      "Processing P24627 (Index: 33 of 42) took 1.59 seconds.\n",
      "Processing P02788 (Index: 34 of 42) took 2.02 seconds.\n",
      "Processing P14632 (Index: 35 of 42) took 1.57 seconds.\n",
      "Processing Q29477 (Index: 36 of 42) took 1.57 seconds.\n",
      "Processing O77698 (Index: 37 of 42) took 1.57 seconds.\n",
      "Processing P14639 (Index: 38 of 42) took 1.58 seconds.\n",
      "Processing P02769 (Index: 39 of 42) took 1.56 seconds.\n",
      "Processing P67976 (Index: 40 of 42) took 1.58 seconds.\n",
      "Processing L8I8G5 (Index: 41 of 42) took 1.62 seconds.\n",
      "Processing P01966 (Index: 42 of 42) took 1.60 seconds.\n"
     ]
    }
   ],
   "source": [
    "processed_peptide_df, processed_reference_df = process_species_data(uniprot_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc438477",
   "metadata": {},
   "source": [
    "A series of merge steps to combine peptide and reference data, exands peptide list if multiple references are mentioned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "fea3054b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Explode the 'evidence_pubmed' column, creating a row for each entry in the list\n",
    "expanded_df = processed_peptide_df.explode('evidence_pubmed')\n",
    "\n",
    "# Remove rows where 'evidence_pubmed' is NaN\n",
    "filtered_df = expanded_df.dropna(subset=['evidence_pubmed'])\n",
    "\n",
    "# Merge the filtered dataframe with the reference dataframe based on 'proteinID' and 'evidence_pubmed'\n",
    "merged_df = pd.merge(filtered_df, processed_reference_df, \n",
    "                     left_on=['proteinID', 'evidence_pubmed'], \n",
    "                     right_on=['proteinID', 'pubmed'], \n",
    "                     how='left')\n",
    "\n",
    "# Extract rows from expanded_df where 'evidence_pubmed' is NaN\n",
    "no_pubmed_df = expanded_df[expanded_df['evidence_pubmed'].isna()]\n",
    "\n",
    "# Merge no_pubmed_df with reference dataframe based on 'proteinID' and 'interval'\n",
    "merged_df_nopubmed_df = pd.merge(no_pubmed_df, processed_reference_df, \n",
    "                                left_on=['proteinID', 'interval'], \n",
    "                                right_on=['proteinID', 'scope_range'], \n",
    "                                how='left')\n",
    "\n",
    "# Filter rows where the interval matches the scope_range and 'pubmed' is NaN\n",
    "matching_interval_df_no_pubmed = merged_df_nopubmed_df[(merged_df_nopubmed_df['interval'] == merged_df_nopubmed_df['scope_range']) &  merged_df_nopubmed_df['pubmed'].isna()]\n",
    "\n",
    "# Combine merged_df and merged_df_nopubmed_df into a single dataframe\n",
    "full_merged_df = pd.concat([merged_df, merged_df_nopubmed_df], ignore_index=True)\n",
    "\n",
    "# Create 'pubmed_id' column based on conditions related to 'evidence_pubmed' and 'pubmed'\n",
    "full_merged_df['pubmed_id'] = full_merged_df.apply(lambda row: row['evidence_pubmed'] if pd.isna(row['pubmed']) else (row['pubmed'] if pd.isna(row['evidence_pubmed']) else (row['evidence_pubmed'] if row['evidence_pubmed'] == row['pubmed'] else None)), axis=1)\n",
    "\n",
    "# Filter rows from full_merged_df where 'pubmed_id' is not NaN\n",
    "reff_merged_df = full_merged_df[full_merged_df['pubmed_id'].notna()].copy()\n",
    "\n",
    "# Add rows from matching_interval_df_no_pubmed to reff_merged_df\n",
    "reff_merged_df = pd.concat([reff_merged_df, matching_interval_df_no_pubmed], ignore_index=True)\n",
    "\n",
    "# Create a unique identifier for each row in both dataframes\n",
    "full_merged_df['uid'] = full_merged_df['proteinID'] + '_' + full_merged_df['interval'].astype(str)\n",
    "matching_interval_df_no_pubmed.loc[:, 'uid'] = matching_interval_df_no_pubmed['proteinID'] + '_' + matching_interval_df_no_pubmed['interval'].astype(str)\n",
    "\n",
    "# Check if rows in full_merged_df exist in matching_interval_df_no_pubmed using the unique identifier\n",
    "full_merged_df['in_matching'] = full_merged_df['uid'].isin(matching_interval_df_no_pubmed['uid'])\n",
    "\n",
    "# Filter rows in full_merged_df where 'pubmed_id' is NaN and aren't present in matching_interval_df_no_pubmed\n",
    "no_reff_merged_df = full_merged_df[(full_merged_df['pubmed_id'].isna()) & (~full_merged_df['in_matching'])].copy()\n",
    "\n",
    "# Drop the temporary columns used for operations\n",
    "no_reff_merged_df.drop(columns=['uid', 'in_matching'], inplace=True)\n",
    "full_merged_df.drop(columns=['uid', 'in_matching'], inplace=True)\n",
    "matching_interval_df_no_pubmed.drop(columns=['uid'], inplace=True)\n",
    "\n",
    "# Suppress the SettingWithCopyWarning that may arise from certain operations\n",
    "pd.options.mode.chained_assignment = None\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e49dc9a3",
   "metadata": {},
   "source": [
    "Fetchess abstract and reference infromation from PubMed, takes ~1 second per row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "a135075b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row 1/235: Successfully fetched details for PubMed ID 10604470\n",
      "Row 2/235: Successfully fetched details for PubMed ID 10604470\n",
      "Row 3/235: Successfully fetched details for PubMed ID 6208535\n",
      "Row 4/235: Successfully fetched details for PubMed ID 6760865\n",
      "Row 5/235: Successfully fetched details for PubMed ID 7955907\n",
      "Row 6/235: Successfully fetched details for PubMed ID 2532366\n",
      "Row 7/235: Successfully fetched details for PubMed ID 8653797\n",
      "Row 8/235: Successfully fetched details for PubMed ID 2162527\n",
      "Row 9/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 10/235: Successfully fetched details for PubMed ID 7984506\n",
      "Row 11/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 12/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 13/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 14/235: Successfully fetched details for PubMed ID 11145122\n",
      "Row 15/235: Successfully fetched details for PubMed ID 7955907\n",
      "Row 16/235: Successfully fetched details for PubMed ID 2532366\n",
      "Row 17/235: Successfully fetched details for PubMed ID 8653797\n",
      "Row 18/235: Successfully fetched details for PubMed ID 2162527\n",
      "Row 19/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 20/235: Successfully fetched details for PubMed ID 7984506\n",
      "Row 21/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 22/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 23/235: Successfully fetched details for PubMed ID 7831500\n",
      "Row 24/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 25/235: Successfully fetched details for PubMed ID 11145122\n",
      "Row 26/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 27/235: Successfully fetched details for PubMed ID 7984506\n",
      "Row 28/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 29/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 30/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 31/235: Successfully fetched details for PubMed ID 11145122\n",
      "Row 32/235: Successfully fetched details for PubMed ID 7955907\n",
      "Row 33/235: Successfully fetched details for PubMed ID 2532366\n",
      "Row 34/235: Successfully fetched details for PubMed ID 8653797\n",
      "Row 35/235: Successfully fetched details for PubMed ID 2162527\n",
      "Row 36/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 37/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 38/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 39/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 40/235: Successfully fetched details for PubMed ID 11145122\n",
      "Row 41/235: Successfully fetched details for PubMed ID 8853410\n",
      "Row 42/235: Successfully fetched details for PubMed ID 8384600\n",
      "Row 43/235: Successfully fetched details for PubMed ID 2528951\n",
      "Row 44/235: Successfully fetched details for PubMed ID 2972874\n",
      "Row 45/235: Successfully fetched details for PubMed ID 9893117\n",
      "Row 46/235: Successfully fetched details for PubMed ID 8351194\n",
      "Row 47/235: Successfully fetched details for PubMed ID 8779891\n",
      "Row 48/235: Successfully fetched details for PubMed ID 7955907\n",
      "Row 49/235: Successfully fetched details for PubMed ID 2532366\n",
      "Row 50/235: Successfully fetched details for PubMed ID 8653797\n",
      "Row 51/235: Successfully fetched details for PubMed ID 2162527\n",
      "Row 52/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 53/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 54/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 55/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 56/235: Successfully fetched details for PubMed ID 11145122\n",
      "Row 57/235: Successfully fetched details for PubMed ID 8653797\n",
      "Row 58/235: Successfully fetched details for PubMed ID 16875975\n",
      "Row 59/235: Successfully fetched details for PubMed ID 1826098\n",
      "Row 60/235: Successfully fetched details for PubMed ID 8384600\n",
      "Row 61/235: Successfully fetched details for PubMed ID 25401746\n",
      "Row 62/235: Successfully fetched details for PubMed ID 18835931\n",
      "Row 63/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 64/235: Successfully fetched details for PubMed ID 22307324\n",
      "Row 65/235: Successfully fetched details for PubMed ID 2532366\n",
      "Row 66/235: Successfully fetched details for PubMed ID 1660465\n",
      "Row 67/235: Successfully fetched details for PubMed ID 21098034\n",
      "Row 68/235: Successfully fetched details for PubMed ID 2162527\n",
      "Row 69/235: Successfully fetched details for PubMed ID 21672517\n",
      "Row 70/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 71/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 72/235: Successfully fetched details for PubMed ID 10880574\n",
      "Row 73/235: Successfully fetched details for PubMed ID 1672777\n",
      "Row 74/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 75/235: Successfully fetched details for PubMed ID 6230082\n",
      "Row 76/235: Successfully fetched details for PubMed ID 9893117\n",
      "Row 77/235: Successfully fetched details for PubMed ID 3838202\n",
      "Row 78/235: Successfully fetched details for PubMed ID 15741263\n",
      "Row 79/235: Successfully fetched details for PubMed ID 7955907\n",
      "Row 80/235: Successfully fetched details for PubMed ID 2532366\n",
      "Row 81/235: Successfully fetched details for PubMed ID 8653797\n",
      "Row 82/235: Successfully fetched details for PubMed ID 2162527\n",
      "Row 83/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 84/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 85/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 86/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 87/235: Successfully fetched details for PubMed ID 11145122\n",
      "Row 88/235: Successfully fetched details for PubMed ID 7955907\n",
      "Row 89/235: Successfully fetched details for PubMed ID 2532366\n",
      "Row 90/235: Successfully fetched details for PubMed ID 8653797\n",
      "Row 91/235: Successfully fetched details for PubMed ID 2162527\n",
      "Row 92/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 93/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 94/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 95/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 96/235: Successfully fetched details for PubMed ID 11145122\n",
      "Row 97/235: Successfully fetched details for PubMed ID 7955907\n",
      "Row 98/235: Successfully fetched details for PubMed ID 2532366\n",
      "Row 99/235: Successfully fetched details for PubMed ID 8653797\n",
      "Row 100/235: Successfully fetched details for PubMed ID 2162527\n",
      "Row 101/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 102/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 103/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 104/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 105/235: Successfully fetched details for PubMed ID 11145122\n",
      "Row 106/235: Successfully fetched details for PubMed ID 7955907\n",
      "Row 107/235: Successfully fetched details for PubMed ID 2532366\n",
      "Row 108/235: Successfully fetched details for PubMed ID 8653797\n",
      "Row 109/235: Successfully fetched details for PubMed ID 2162527\n",
      "Row 110/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 111/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 112/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 113/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 114/235: Successfully fetched details for PubMed ID 11145122\n",
      "Row 115/235: Successfully fetched details for PubMed ID 7955907\n",
      "Row 116/235: Successfully fetched details for PubMed ID 2532366\n",
      "Row 117/235: Successfully fetched details for PubMed ID 8653797\n",
      "Row 118/235: Successfully fetched details for PubMed ID 2162527\n",
      "Row 119/235: Successfully fetched details for PubMed ID 7720651\n",
      "Row 120/235: Successfully fetched details for PubMed ID 7595132\n",
      "Row 121/235: Successfully fetched details for PubMed ID 8087923\n",
      "Row 122/235: Successfully fetched details for PubMed ID 2825692\n",
      "Row 123/235: Successfully fetched details for PubMed ID 11145122\n",
      "Row 124/235: Successfully fetched details for PubMed ID 2909530\n",
      "Row 125/235: Successfully fetched details for PubMed ID 8376604\n",
      "Row 126/235: Successfully fetched details for PubMed ID 16333141\n",
      "Row 127/235: Successfully fetched details for PubMed ID 15833747\n",
      "Row 128/235: Successfully fetched details for PubMed ID 9059512\n",
      "Row 129/235: Successfully fetched details for PubMed ID 10432298\n",
      "Row 130/235: Successfully fetched details for PubMed ID 19615750\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row 131/235: Successfully fetched details for PubMed ID 25664854\n",
      "Row 132/235: Successfully fetched details for PubMed ID 15187149\n",
      "Row 133/235: Successfully fetched details for PubMed ID 22879591\n",
      "Row 134/235: Successfully fetched details for PubMed ID 29615476\n",
      "Row 135/235: Successfully fetched details for PubMed ID 29615476\n",
      "Row 136/235: Successfully fetched details for PubMed ID 12693969\n",
      "Row 137/235: Successfully fetched details for PubMed ID 1369293\n",
      "Row 138/235: Successfully fetched details for PubMed ID 1369293\n",
      "Row 139/235: Successfully fetched details for PubMed ID 1369293\n",
      "Row 140/235: Successfully fetched details for PubMed ID 12665801\n",
      "Row 141/235: Successfully fetched details for PubMed ID 7493937\n",
      "Row 142/235: Successfully fetched details for PubMed ID 7592911\n",
      "Row 143/235: Successfully fetched details for PubMed ID 7529412\n",
      "Row 144/235: Successfully fetched details for PubMed ID 8681941\n",
      "Row 145/235: Successfully fetched details for PubMed ID 18818205\n",
      "Row 146/235: Successfully fetched details for PubMed ID 7529412\n",
      "Row 147/235: Successfully fetched details for PubMed ID 15778390\n",
      "Row 148/235: Successfully fetched details for PubMed ID 16637646\n",
      "Row 149/235: Successfully fetched details for PubMed ID 9736536\n",
      "Row 150/235: Successfully fetched details for PubMed ID 33060695\n",
      "Row 151/235: Successfully fetched details for PubMed ID 10417311\n",
      "Row 152/235: Successfully fetched details for PubMed ID 22879591\n",
      "Row 153/235: Successfully fetched details for PubMed ID 34708076\n",
      "Row 154/235: Successfully fetched details for PubMed ID 29133814\n",
      "Row 155/235: Successfully fetched details for PubMed ID 8681941\n",
      "Row 156/235: Successfully fetched details for PubMed ID 32753597\n",
      "Row 157/235: Successfully fetched details for PubMed ID 18818205\n",
      "Row 158/235: Successfully fetched details for PubMed ID 16637646\n",
      "Row 159/235: Successfully fetched details for PubMed ID 9736536\n",
      "Row 160/235: Successfully fetched details for PubMed ID 14978112\n",
      "Row 161/235: Successfully fetched details for PubMed ID 22879591\n",
      "Row 162/235: Successfully fetched details for PubMed ID 18818205\n",
      "Row 163/235: Successfully fetched details for PubMed ID 16637646\n",
      "Row 164/235: Successfully fetched details for PubMed ID 9736536\n",
      "Row 165/235: Successfully fetched details for PubMed ID 14978112\n",
      "Row 166/235: Successfully fetched details for PubMed ID 22879591\n",
      "Row 167/235: Successfully fetched details for PubMed ID 18818205\n",
      "Row 168/235: Successfully fetched details for PubMed ID 16637646\n",
      "Row 169/235: Successfully fetched details for PubMed ID 9736536\n",
      "Row 170/235: Successfully fetched details for PubMed ID 14978112\n",
      "Row 171/235: Successfully fetched details for PubMed ID 22879591\n",
      "Row 172/235: Successfully fetched details for PubMed ID 18818205\n",
      "Row 173/235: Successfully fetched details for PubMed ID 16637646\n",
      "Row 174/235: Successfully fetched details for PubMed ID 9736536\n",
      "Row 175/235: Successfully fetched details for PubMed ID 14978112\n",
      "Row 176/235: Successfully fetched details for PubMed ID 22879591\n",
      "Row 177/235: Successfully fetched details for PubMed ID 18818205\n",
      "Row 178/235: Successfully fetched details for PubMed ID 16637646\n",
      "Row 179/235: Successfully fetched details for PubMed ID 9736536\n",
      "Row 180/235: Successfully fetched details for PubMed ID 14978112\n",
      "Row 181/235: Successfully fetched details for PubMed ID 22879591\n",
      "Row 182/235: Successfully fetched details for PubMed ID 18818205\n",
      "Row 183/235: Successfully fetched details for PubMed ID 16637646\n",
      "Row 184/235: Successfully fetched details for PubMed ID 9736536\n",
      "Row 185/235: Successfully fetched details for PubMed ID 14978112\n",
      "Row 186/235: Successfully fetched details for PubMed ID 22879591\n",
      "Row 187/235: Successfully fetched details for PubMed ID 7057924\n",
      "Row 188/235: Successfully fetched details for PubMed ID 656131\n",
      "Row 189/235: Successfully fetched details for PubMed ID 7057924\n",
      "Row 190/235: Successfully fetched details for PubMed ID 656131\n",
      "Row 191/235: Successfully fetched details for PubMed ID 7057924\n",
      "Row 192/235: Successfully fetched details for PubMed ID 2982830\n",
      "Row 193/235: Successfully fetched details for PubMed ID 7057924\n",
      "Row 194/235: Successfully fetched details for PubMed ID 7057924\n",
      "Row 195/235: Successfully fetched details for PubMed ID 5760861\n",
      "Row 196/235: Successfully fetched details for PubMed ID 13591312\n",
      "Row 197/235: Successfully fetched details for PubMed ID 12526812\n",
      "Row 198/235: Successfully fetched details for PubMed ID 28453791\n",
      "Row 199/235: Successfully fetched details for PubMed ID 12554744\n",
      "Row 200/235: Successfully fetched details for PubMed ID 10605628\n",
      "Row 201/235: Successfully fetched details for PubMed ID 12554744\n",
      "Row 202/235: Successfully fetched details for PubMed ID 10605628\n",
      "Row 203/235: Successfully fetched details for PubMed ID 11946536\n",
      "Row 204/235: Successfully fetched details for PubMed ID 12626323\n",
      "Row 205/235: Successfully fetched details for PubMed ID 10605628\n",
      "Row 206/235: Successfully fetched details for PubMed ID 22037645\n",
      "Row 207/235: Successfully fetched details for PubMed ID 12554744\n",
      "Row 208/235: Successfully fetched details for PubMed ID 2753890\n",
      "Row 209/235: Successfully fetched details for PubMed ID 14719035\n",
      "Row 210/235: Successfully fetched details for PubMed ID 10605628\n",
      "Row 211/235: Successfully fetched details for PubMed ID 2753890\n",
      "Row 212/235: Successfully fetched details for PubMed ID 2753890\n",
      "Row 213/235: Successfully fetched details for PubMed ID 10605628\n",
      "Row 214/235: Successfully fetched details for PubMed ID 10322410\n",
      "Row 215/235: Successfully fetched details for PubMed ID 12554744\n",
      "Row 216/235: Successfully fetched details for PubMed ID 14719035\n",
      "Row 217/235: Successfully fetched details for PubMed ID 10438732\n",
      "Row 218/235: Successfully fetched details for PubMed ID 10438732\n",
      "Row 219/235: Successfully fetched details for PubMed ID 3654650\n",
      "Row 220/235: Successfully fetched details for PubMed ID 4056036\n",
      "Row 221/235: Successfully fetched details for PubMed ID 3317417\n",
      "Row 222/235: Successfully fetched details for PubMed ID 2091067\n",
      "Row 223/235: Successfully fetched details for PubMed ID 3035556\n",
      "Row 224/235: Successfully fetched details for PubMed ID 3342889\n",
      "Row 225/235: Successfully fetched details for PubMed ID 3202875\n",
      "Row 226/235: Successfully fetched details for PubMed ID 2587421\n",
      "Row 227/235: Successfully fetched details for PubMed ID 631332\n",
      "Row 228/235: Successfully fetched details for PubMed ID 6812220\n",
      "Row 229/235: Successfully fetched details for PubMed ID 6745415\n",
      "Row 230/235: Successfully fetched details for PubMed ID 6366786\n",
      "Row 231/235: Successfully fetched details for PubMed ID 14426955\n",
      "Row 232/235: Successfully fetched details for PubMed ID 5921183\n",
      "Row 233/235: Successfully fetched details for PubMed ID 5822140\n",
      "Row 234/235: Successfully fetched details for PubMed ID 7628632\n",
      "Row 235/235: Successfully fetched details for PubMed ID 36459277\n"
     ]
    }
   ],
   "source": [
    "# Initialize an empty DataFrame to store details fetched from the 'fetch_details' function\n",
    "pubmed_details = pd.DataFrame()\n",
    "\n",
    "# Calculate the total number of rows in the reff_merged_df for progress tracking\n",
    "total_rows = len(reff_merged_df)\n",
    "\n",
    "# Apply the 'fetch_details' function to each row of the reff_merged_df\n",
    "# The results are stored in a Series where each entry is a dictionary of details fetched for that row\n",
    "results = reff_merged_df.apply(lambda row: fetch_details(row, row.name + 1, total_rows), axis=1)\n",
    "\n",
    "# Extract details from the results and assign them to appropriate columns in the pubmed_details DataFrame\n",
    "pubmed_details['pubmed_id'] = results.apply(lambda x: x['pubmed_id'] if x else None)   # Extract 'pubmed_id' from the result dictionary\n",
    "pubmed_details['abstract'] = results.apply(lambda x: x['abstract'] if x else None)     # Extract 'abstract' from the result dictionary\n",
    "pubmed_details['title'] = results.apply(lambda x: x['title'] if x else None)           # Extract 'title' from the result dictionary\n",
    "pubmed_details['authors'] = results.apply(lambda x: x['authors'] if x else None)       # Extract 'authors' from the result dictionary\n",
    "\n",
    "# Check if 'doi' exists in the result dictionary before extracting to avoid KeyError\n",
    "pubmed_details['doi'] = results.apply(lambda x: x['doi'] if x and 'doi' in x else None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8e088d4",
   "metadata": {},
   "source": [
    "The extracted pubmed data can be explored below, where pubmedID was nan, errounus references can be extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "8922200d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2af1617d81034553aa1099d31bd316f6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': False, 'defa…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pubmed_details_qgrid = qgrid.show_grid(pubmed_details, show_toolbar=True, grid_options={'forceFitColumns': False})\n",
    "pubmed_details_qgrid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "9088a885",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Information for DataFrame: no_reff_merged_final_df\n",
      "----------------------------------------\n",
      "Numbers of peptides: 63\n",
      "Shape: (63, 7)\n",
      "Columns: ['proteinID', 'chain', 'interval', 'peptide', 'description', 'associated_function', 'non_associated_function']\n",
      "\n",
      "Missing Values Count:\n",
      "proteinID                   0\n",
      "chain                       0\n",
      "interval                    0\n",
      "peptide                     0\n",
      "description                 0\n",
      "associated_function         0\n",
      "non_associated_function    20\n",
      "dtype: int64\n",
      "\n",
      "========================================\n",
      "\n",
      "Information for DataFrame: reff_merged_final_df\n",
      "----------------------------------------\n",
      "Numbers of peptides: 235\n",
      "Shape: (235, 12)\n",
      "Columns: ['proteinID', 'chain', 'interval', 'peptide', 'description', 'associated_function', 'non_associated_function', 'title', 'authors', 'doi', 'pubmed_id', 'abstract']\n",
      "\n",
      "Missing Values Count:\n",
      "proteinID                    0\n",
      "chain                        0\n",
      "interval                     0\n",
      "peptide                      0\n",
      "description                  0\n",
      "associated_function          0\n",
      "non_associated_function    208\n",
      "title                        0\n",
      "authors                      0\n",
      "doi                          3\n",
      "pubmed_id                    1\n",
      "abstract                    15\n",
      "dtype: int64\n",
      "\n",
      "========================================\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Remove duplicate entries from the pubmed_details DataFrame based on the 'pubmed_id' column\n",
    "pubmed_details_unique = pubmed_details.drop_duplicates(subset='pubmed_id')\n",
    "\n",
    "# Merge the reff_merged_df with the unique pubmed_details based on 'pubmed_id'\n",
    "# If there's a name conflict between the columns of the two dataframes being merged, '_pubmed' is added as a suffix to columns from pubmed_details_unique\n",
    "reff_merged_final_df = pd.merge(reff_merged_df, pubmed_details_unique, on='pubmed_id', how='left', suffixes=('', '_pubmed'))\n",
    "\n",
    "# For rows with missing values in the 'title', 'doi', and 'authors' columns of reff_merged_final_df, \n",
    "# replace them with corresponding values from columns 'title_pubmed', 'doi_pubmed', and 'authors_pubmed'\n",
    "reff_merged_final_df['title'] = reff_merged_final_df.apply(lambda row: row['title'] if pd.notnull(row['title']) else row['title_pubmed'], axis=1)\n",
    "reff_merged_final_df['doi'] = reff_merged_final_df.apply(lambda row: row['doi'] if pd.notnull(row['doi']) else row['doi_pubmed'], axis=1)\n",
    "reff_merged_final_df['authors'] = reff_merged_final_df.apply(lambda row: row['authors'] if pd.notnull(row['authors']) else row['authors_pubmed'], axis=1)\n",
    "\n",
    "# Drop columns that are no longer needed from reff_merged_final_df\n",
    "reff_merged_final_df = reff_merged_final_df.drop(columns=['title_pubmed', 'doi_pubmed', 'authors_pubmed', 'pubmed', 'feature_evidence', 'evidence_to_pubmed', 'evidence_pubmed', 'comment_evidence', 'reference key', 'scope_range'])\n",
    "\n",
    "# Drop columns that are not needed from no_reff_merged_df\n",
    "no_reff_merged_final_df = no_reff_merged_df.drop(columns=['pubmed', 'feature_evidence', 'evidence_to_pubmed', 'evidence_pubmed', 'comment_evidence', 'reference key', 'title', 'authors', 'pubmed', 'doi', 'scope_range', 'pubmed_id'])\n",
    "\n",
    "print_critical_info(no_reff_merged_final_df, \"no_reff_merged_final_df\")\n",
    "print_critical_info(reff_merged_final_df, \"reff_merged_final_df\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "623d2f9d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID</th>\n",
       "      <th>chain</th>\n",
       "      <th>interval</th>\n",
       "      <th>peptide</th>\n",
       "      <th>description</th>\n",
       "      <th>associated_function</th>\n",
       "      <th>non_associated_function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>Q9UBU3</td>\n",
       "      <td></td>\n",
       "      <td>76-98</td>\n",
       "      <td>FNAPFDVGIKLSGVQYQQHSQAL</td>\n",
       "      <td>Obestatin</td>\n",
       "      <td>May be the ligand for GPR39. May have an appet...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>P01009</td>\n",
       "      <td>Alpha-1-antitrypsin</td>\n",
       "      <td>375-418</td>\n",
       "      <td>MFLEAIPMSIPPEVKFNKPFVFLMIEQNTKSPLFMGKVVNPTQK</td>\n",
       "      <td>Short peptide from AAT</td>\n",
       "      <td>Reversible chymotrypsin inhibitor. It also inh...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>P30990</td>\n",
       "      <td>Large neuromedin N</td>\n",
       "      <td>144-148</td>\n",
       "      <td>IPYIL</td>\n",
       "      <td>Neuromedin N</td>\n",
       "      <td></td>\n",
       "      <td>[Neurotensin may play an endocrine or paracrin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>P30990</td>\n",
       "      <td>Large neuromedin N</td>\n",
       "      <td>166-170</td>\n",
       "      <td>DSYYY</td>\n",
       "      <td>Tail peptide</td>\n",
       "      <td></td>\n",
       "      <td>[Neurotensin may play an endocrine or paracrin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>P01282</td>\n",
       "      <td></td>\n",
       "      <td>81-107</td>\n",
       "      <td>HADGVFTSDFSKLLGQLSAKKYLESLM</td>\n",
       "      <td>Intestinal peptide PHM-27</td>\n",
       "      <td></td>\n",
       "      <td>[VIP causes vasodilation, lowers arterial bloo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>P28799</td>\n",
       "      <td>Progranulin</td>\n",
       "      <td>123-179</td>\n",
       "      <td>AIQCPDSQFECPDFSTCCVMVDGSWGCCPMPQASCCEDRVHCCPHG...</td>\n",
       "      <td>Granulin-2</td>\n",
       "      <td></td>\n",
       "      <td>[Secreted protein that acts as a key regulator...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>P28799</td>\n",
       "      <td>Progranulin</td>\n",
       "      <td>281-336</td>\n",
       "      <td>DVKCDMEVSCPDGYTCCRLQSGAWGCCPFTQAVCCEDHIHCCPAGF...</td>\n",
       "      <td>Granulin-4</td>\n",
       "      <td>Promotes proliferation of the epithelial cell ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>P28799</td>\n",
       "      <td>Progranulin</td>\n",
       "      <td>364-417</td>\n",
       "      <td>VPCDNVSSCPSSDTCCQLTSGEWGCCPIPEAVCCSDHQHCCPQGYT...</td>\n",
       "      <td>Granulin-5</td>\n",
       "      <td></td>\n",
       "      <td>[Secreted protein that acts as a key regulator...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>P28799</td>\n",
       "      <td>Progranulin</td>\n",
       "      <td>442-496</td>\n",
       "      <td>IGCDQHTSCPVGQTCCPSLGGSWACCQLPHAVCCEDRQHCCPAGYT...</td>\n",
       "      <td>Granulin-6</td>\n",
       "      <td></td>\n",
       "      <td>[Secreted protein that acts as a key regulator...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>P01275</td>\n",
       "      <td></td>\n",
       "      <td>21-50</td>\n",
       "      <td>RSLQDTEEKSRSFSASQADPLSDPDQMNED</td>\n",
       "      <td>Glicentin-related polypeptide</td>\n",
       "      <td></td>\n",
       "      <td>[Plays a key role in glucose metabolism and ho...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    proteinID                chain interval  \\\n",
       "218    Q9UBU3                         76-98   \n",
       "219    P01009  Alpha-1-antitrypsin  375-418   \n",
       "220    P30990   Large neuromedin N  144-148   \n",
       "221    P30990   Large neuromedin N  166-170   \n",
       "223    P01282                        81-107   \n",
       "..        ...                  ...      ...   \n",
       "293    P28799          Progranulin  123-179   \n",
       "294    P28799          Progranulin  281-336   \n",
       "295    P28799          Progranulin  364-417   \n",
       "296    P28799          Progranulin  442-496   \n",
       "297    P01275                         21-50   \n",
       "\n",
       "                                               peptide  \\\n",
       "218                            FNAPFDVGIKLSGVQYQQHSQAL   \n",
       "219       MFLEAIPMSIPPEVKFNKPFVFLMIEQNTKSPLFMGKVVNPTQK   \n",
       "220                                              IPYIL   \n",
       "221                                              DSYYY   \n",
       "223                        HADGVFTSDFSKLLGQLSAKKYLESLM   \n",
       "..                                                 ...   \n",
       "293  AIQCPDSQFECPDFSTCCVMVDGSWGCCPMPQASCCEDRVHCCPHG...   \n",
       "294  DVKCDMEVSCPDGYTCCRLQSGAWGCCPFTQAVCCEDHIHCCPAGF...   \n",
       "295  VPCDNVSSCPSSDTCCQLTSGEWGCCPIPEAVCCSDHQHCCPQGYT...   \n",
       "296  IGCDQHTSCPVGQTCCPSLGGSWACCQLPHAVCCEDRQHCCPAGYT...   \n",
       "297                     RSLQDTEEKSRSFSASQADPLSDPDQMNED   \n",
       "\n",
       "                       description  \\\n",
       "218                      Obestatin   \n",
       "219         Short peptide from AAT   \n",
       "220                   Neuromedin N   \n",
       "221                   Tail peptide   \n",
       "223      Intestinal peptide PHM-27   \n",
       "..                             ...   \n",
       "293                     Granulin-2   \n",
       "294                     Granulin-4   \n",
       "295                     Granulin-5   \n",
       "296                     Granulin-6   \n",
       "297  Glicentin-related polypeptide   \n",
       "\n",
       "                                   associated_function  \\\n",
       "218  May be the ligand for GPR39. May have an appet...   \n",
       "219  Reversible chymotrypsin inhibitor. It also inh...   \n",
       "220                                                      \n",
       "221                                                      \n",
       "223                                                      \n",
       "..                                                 ...   \n",
       "293                                                      \n",
       "294  Promotes proliferation of the epithelial cell ...   \n",
       "295                                                      \n",
       "296                                                      \n",
       "297                                                      \n",
       "\n",
       "                               non_associated_function  \n",
       "218                                                NaN  \n",
       "219                                                NaN  \n",
       "220  [Neurotensin may play an endocrine or paracrin...  \n",
       "221  [Neurotensin may play an endocrine or paracrin...  \n",
       "223  [VIP causes vasodilation, lowers arterial bloo...  \n",
       "..                                                 ...  \n",
       "293  [Secreted protein that acts as a key regulator...  \n",
       "294                                                NaN  \n",
       "295  [Secreted protein that acts as a key regulator...  \n",
       "296  [Secreted protein that acts as a key regulator...  \n",
       "297  [Plays a key role in glucose metabolism and ho...  \n",
       "\n",
       "[63 rows x 7 columns]"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_reff_merged_final_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e240122a",
   "metadata": {},
   "source": [
    "Extracteds asccociated function from non_asscoiated by referencing pubmed ID and description column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "29d81d07",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In reff_merged dataframe, 6 associated_functions were assigned.\n",
      "In no_reff_merged dataframe, 3 associated_functions were assigned.\n"
     ]
    }
   ],
   "source": [
    "# Calculate initial counts of empty 'associated_function' in both dataframes\n",
    "initial_empty_reff = (reff_merged_final_df['associated_function'] == '').sum()\n",
    "initial_empty_no_reff = (no_reff_merged_final_df['associated_function'] == '').sum()\n",
    "\n",
    "# Apply the update_associated_function method to the 'associated_function' column of both dataframes\n",
    "reff_merged_final_df['associated_function'] = reff_merged_final_df.apply(update_associated_function, axis=1)\n",
    "no_reff_merged_final_df['associated_function'] = no_reff_merged_final_df.apply(update_associated_function, axis=1)\n",
    "\n",
    "# Calculate the change in the number of empty 'associated_function' entries in both dataframes\n",
    "change_reff = initial_empty_reff - (reff_merged_final_df['associated_function'] == '').sum()\n",
    "change_no_reff = initial_empty_no_reff - (no_reff_merged_final_df['associated_function'] == '').sum()\n",
    "\n",
    "# Print the change\n",
    "print(f\"In reff_merged dataframe, {change_reff} associated_functions were assigned.\")\n",
    "print(f\"In no_reff_merged dataframe, {change_no_reff} associated_functions were assigned.\")\n",
    "\n",
    "# Set 'non_associated_function' to NaN for rows where 'associated_function' is not NaN\n",
    "reff_merged_final_df.loc[reff_merged_final_df['associated_function'].notna() & (reff_merged_final_df['associated_function'] != \"\"), 'non_associated_function'] = np.nan\n",
    "no_reff_merged_final_df.loc[no_reff_merged_final_df['associated_function'].notna() & (no_reff_merged_final_df['associated_function'] != \"\"), 'non_associated_function'] = np.nan"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7186cb96",
   "metadata": {},
   "source": [
    "# Exports/Imports data to CSV"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebe75afb",
   "metadata": {},
   "source": [
    "Exports real protein list to CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "0803fa63",
   "metadata": {},
   "outputs": [],
   "source": [
    "no_reff_merged_final_df.to_csv('data/no_reff_merged_final_df.csv')\n",
    "reff_merged_final_df.to_csv('data/reff_merged_final_df.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebc1a20d",
   "metadata": {},
   "source": [
    "Exports test rotein list to CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "dcb79d3a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "uni_no_reff_merged_final_df.to_csv('data/uni_no_reff_merged_final_df.csv')\n",
    "uni_reff_merged_final_df.to_csv('data/uni_reff_merged_final_df.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64a12ad7",
   "metadata": {},
   "source": [
    "Import test dataframes from CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a0a49903",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "(unicode error) 'unicodeescape' codec can't decode bytes in position 4-5: truncated \\uXXXX escape (Temp/ipykernel_20236/3315866054.py, line 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"C:\\Users\\kuhfeldr\\AppData\\Local\\Temp/ipykernel_20236/3315866054.py\"\u001b[1;36m, line \u001b[1;32m2\u001b[0m\n\u001b[1;33m    uni_no_reff_merged_final_df = pd.read_csv('data\\uni_no_reff_merged_final_df.csv', index_col=0)\u001b[0m\n\u001b[1;37m                                                                                    ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m (unicode error) 'unicodeescape' codec can't decode bytes in position 4-5: truncated \\uXXXX escape\n"
     ]
    }
   ],
   "source": [
    "uni_reff_merged_final_df = pd.read_csv('data/uni_reff_merged_final_df.csv', index_col=0)\n",
    "uni_no_reff_merged_final_df = pd.read_csv('data/uni_no_reff_merged_final_df.csv', index_col=0)\n",
    "uni_reff_merged_final_df['pubmed_id'] = uni_reff_merged_final_df['pubmed_id'].astype('Int64')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a833850",
   "metadata": {},
   "source": [
    "Import real dataframes from CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c55f043b",
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "[Errno 22] Invalid argument: 'data\\reff_merged_final_df.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_20236/2409851291.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mreff_merged_final_df\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'data\\reff_merged_final_df.csv'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindex_col\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[0mno_reff_merged_final_df\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'data\\no_reff_merged_final_df.csv'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindex_col\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;31m#contverts pubmed ID from float to int, deals with NAs\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0mreff_merged_final_df\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'pubmed_id'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mreff_merged_final_df\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'pubmed_id'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwhere\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mnotna\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mreff_merged_final_df\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'pubmed_id'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mNA\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0mreff_merged_final_df\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'pubmed_id'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mreff_merged_final_df\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'pubmed_id'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Int64'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\util\\_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    309\u001b[0m                     \u001b[0mstacklevel\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mstacklevel\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    310\u001b[0m                 )\n\u001b[1;32m--> 311\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    312\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    313\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\io\\parsers\\readers.py\u001b[0m in \u001b[0;36mread_csv\u001b[1;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[0;32m    584\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    585\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 586\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    587\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    588\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\io\\parsers\\readers.py\u001b[0m in \u001b[0;36m_read\u001b[1;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[0;32m    480\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    481\u001b[0m     \u001b[1;31m# Create the parser.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 482\u001b[1;33m     \u001b[0mparser\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    483\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    484\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\io\\parsers\\readers.py\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[0;32m    809\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"has_index_names\"\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"has_index_names\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    810\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 811\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    812\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    813\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\io\\parsers\\readers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[1;34m(self, engine)\u001b[0m\n\u001b[0;32m   1038\u001b[0m             )\n\u001b[0;32m   1039\u001b[0m         \u001b[1;31m# error: Too many arguments for \"ParserBase\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1040\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mmapping\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mengine\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[1;33m)\u001b[0m  \u001b[1;31m# type: ignore[call-arg]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1041\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1042\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_failover_to_python\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\io\\parsers\\c_parser_wrapper.py\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, src, **kwds)\u001b[0m\n\u001b[0;32m     49\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     50\u001b[0m         \u001b[1;31m# open handles\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 51\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_open_handles\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     52\u001b[0m         \u001b[1;32massert\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhandles\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     53\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\io\\parsers\\base_parser.py\u001b[0m in \u001b[0;36m_open_handles\u001b[1;34m(self, src, kwds)\u001b[0m\n\u001b[0;32m    220\u001b[0m         \u001b[0mLet\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mreaders\u001b[0m \u001b[0mopen\u001b[0m \u001b[0mIOHandles\u001b[0m \u001b[0mafter\u001b[0m \u001b[0mthey\u001b[0m \u001b[0mare\u001b[0m \u001b[0mdone\u001b[0m \u001b[1;32mwith\u001b[0m \u001b[0mtheir\u001b[0m \u001b[0mpotential\u001b[0m \u001b[0mraises\u001b[0m\u001b[1;33m.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    221\u001b[0m         \"\"\"\n\u001b[1;32m--> 222\u001b[1;33m         self.handles = get_handle(\n\u001b[0m\u001b[0;32m    223\u001b[0m             \u001b[0msrc\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    224\u001b[0m             \u001b[1;34m\"r\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\io\\common.py\u001b[0m in \u001b[0;36mget_handle\u001b[1;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[0;32m    700\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mioargs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mencoding\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;34m\"b\"\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mioargs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    701\u001b[0m             \u001b[1;31m# Encoding\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 702\u001b[1;33m             handle = open(\n\u001b[0m\u001b[0;32m    703\u001b[0m                 \u001b[0mhandle\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    704\u001b[0m                 \u001b[0mioargs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mOSError\u001b[0m: [Errno 22] Invalid argument: 'data\\reff_merged_final_df.csv'"
     ]
    }
   ],
   "source": [
    "reff_merged_final_df = pd.read_csv('data/reff_merged_final_df.csv', index_col=0)\n",
    "no_reff_merged_final_df = pd.read_csv('data/no_reff_merged_final_df.csv', index_col=0)\n",
    "#contverts pubmed ID from float to int, deals with NAs\n",
    "reff_merged_final_df['pubmed_id'] = reff_merged_final_df['pubmed_id'].where(pd.notna(reff_merged_final_df['pubmed_id']), pd.NA)\n",
    "reff_merged_final_df['pubmed_id'] = reff_merged_final_df['pubmed_id'].astype('Int64')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4b4152c",
   "metadata": {},
   "source": [
    "# Explores Organized data / make corrections to associated function"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3f618f9",
   "metadata": {},
   "source": [
    "Prints key info on dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "69401201",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Information for DataFrame: no_reff_merged_final_df\n",
      "----------------------------------------\n",
      "Numbers of peptides: 63\n",
      "Shape: (63, 7)\n",
      "Columns: ['proteinID', 'chain', 'interval', 'peptide', 'description', 'associated_function', 'non_associated_function']\n",
      "\n",
      "Missing Values Count:\n",
      "proteinID                   0\n",
      "chain                      28\n",
      "interval                    0\n",
      "peptide                     0\n",
      "description                 0\n",
      "associated_function        41\n",
      "non_associated_function    23\n",
      "dtype: int64\n",
      "\n",
      "========================================\n",
      "\n",
      "Information for DataFrame: reff_merged_final_df\n",
      "----------------------------------------\n",
      "Numbers of peptides: 235\n",
      "Shape: (235, 12)\n",
      "Columns: ['proteinID', 'chain', 'interval', 'peptide', 'description', 'associated_function', 'non_associated_function', 'title', 'authors', 'doi', 'pubmed_id', 'abstract']\n",
      "\n",
      "Missing Values Count:\n",
      "proteinID                    0\n",
      "chain                       92\n",
      "interval                     0\n",
      "peptide                      0\n",
      "description                  0\n",
      "associated_function         21\n",
      "non_associated_function    214\n",
      "title                        0\n",
      "authors                      0\n",
      "doi                          3\n",
      "pubmed_id                    1\n",
      "abstract                    15\n",
      "dtype: int64\n",
      "\n",
      "========================================\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Print critical information about both DataFrames\n",
    "print_critical_info(no_reff_merged_final_df, \"data/no_reff_merged_final_df\")\n",
    "print_critical_info(reff_merged_final_df, \"data/reff_merged_final_df\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f49a3bc5",
   "metadata": {},
   "source": [
    "### Explores dataframe with references of real data\n",
    "\n",
    "### Edit the dataframe to added the associated functions from non_asscoiated functions that were not parsed:  \n",
    "    1) Copy the asscoiated line line from non_asscociated_function to asscoiated_function.  \n",
    "            Example: Casocidin-1 in description is Casocidin-I in function so this has to be manually parsed.     \n",
    "    2) Delete the non_asscociated_function leaving this blank not nan to indicate a manual edit.      \n",
    "    3) If there is no associated_function the insert \"indeterminable\" into the associated function column and leave the            non_asscoiated text. \n",
    "    4) To save changes run the next line of code with \".get_changed_df()\" in it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "352c81f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "403204b648c24e3a91b5e30fd65a29a5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': False, 'defa…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#uni_reff_merged_final_df_qgrid = qgrid.show_grid(uni_reff_merged_final_df, show_toolbar=True, grid_options={'forceFitColumns': False})\n",
    "#uni_reff_merged_final_df_qgrid\n",
    "\n",
    "reff_merged_final_df_qgrid = qgrid.show_grid(reff_merged_final_df, show_toolbar=True, grid_options={'forceFitColumns': False})\n",
    "reff_merged_final_df_qgrid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "aa7b37e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "reff_merged_final_me_df = reff_merged_final_df_qgrid.get_changed_df()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "254e9a51",
   "metadata": {},
   "source": [
    "### Explores dataframe w/o references of real data\n",
    "\n",
    "### Edit the dataframe to added the associated functions from non_asscoiated functions that were not parsed:  \n",
    "    1) Copy the asscoiated line line from non_asscociated_function to asscoiated_function.  \n",
    "            Example: Casocidin-1 in description is Casocidin-I in function so this has to be manually parsed.     \n",
    "    2) Delete the non_asscociated_function leaving this blank not nan to indicate a manual edit.      \n",
    "    3) If there is no associated_function the insert \"indeterminable\" into the associated function column and leave the         non_asscoiated text. \n",
    "    4) To save changes run the next line of code with \".get_changed_df()\" in it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "a5edfe5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dd010ffbee4c4de4a0446c57b35bd6ca",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': False, 'defa…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#uni_no_reff_merged_final_df_qgrid = qgrid.show_grid(uni_no_reff_merged_final_df, show_toolbar=True, grid_options={'forceFitColumns': False})\n",
    "#uni_no_reff_merged_final_df_qgrid\n",
    "\n",
    "no_reff_merged_final_df_qgrid = qgrid.show_grid(no_reff_merged_final_df, show_toolbar=True, grid_options={'forceFitColumns': False})\n",
    "no_reff_merged_final_df_qgrid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "17780d5f",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID</th>\n",
       "      <th>chain</th>\n",
       "      <th>interval</th>\n",
       "      <th>peptide</th>\n",
       "      <th>description</th>\n",
       "      <th>associated_function</th>\n",
       "      <th>non_associated_function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>Q9UBU3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>76-98</td>\n",
       "      <td>FNAPFDVGIKLSGVQYQQHSQAL</td>\n",
       "      <td>Obestatin</td>\n",
       "      <td>May be the ligand for GPR39. May have an appet...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>P01009</td>\n",
       "      <td>Alpha-1-antitrypsin</td>\n",
       "      <td>375-418</td>\n",
       "      <td>MFLEAIPMSIPPEVKFNKPFVFLMIEQNTKSPLFMGKVVNPTQK</td>\n",
       "      <td>Short peptide from AAT</td>\n",
       "      <td>Reversible chymotrypsin inhibitor. It also inh...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>P30990</td>\n",
       "      <td>Large neuromedin N</td>\n",
       "      <td>144-148</td>\n",
       "      <td>IPYIL</td>\n",
       "      <td>Neuromedin N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Neurotensin may play an endocrine or paracri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>P30990</td>\n",
       "      <td>Large neuromedin N</td>\n",
       "      <td>166-170</td>\n",
       "      <td>DSYYY</td>\n",
       "      <td>Tail peptide</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Neurotensin may play an endocrine or paracri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>P01282</td>\n",
       "      <td>NaN</td>\n",
       "      <td>81-107</td>\n",
       "      <td>HADGVFTSDFSKLLGQLSAKKYLESLM</td>\n",
       "      <td>Intestinal peptide PHM-27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['VIP causes vasodilation, lowers arterial blo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>P28799</td>\n",
       "      <td>Progranulin</td>\n",
       "      <td>123-179</td>\n",
       "      <td>AIQCPDSQFECPDFSTCCVMVDGSWGCCPMPQASCCEDRVHCCPHG...</td>\n",
       "      <td>Granulin-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Secreted protein that acts as a key regulato...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>P28799</td>\n",
       "      <td>Progranulin</td>\n",
       "      <td>281-336</td>\n",
       "      <td>DVKCDMEVSCPDGYTCCRLQSGAWGCCPFTQAVCCEDHIHCCPAGF...</td>\n",
       "      <td>Granulin-4</td>\n",
       "      <td>Promotes proliferation of the epithelial cell ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>P28799</td>\n",
       "      <td>Progranulin</td>\n",
       "      <td>364-417</td>\n",
       "      <td>VPCDNVSSCPSSDTCCQLTSGEWGCCPIPEAVCCSDHQHCCPQGYT...</td>\n",
       "      <td>Granulin-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Secreted protein that acts as a key regulato...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>P28799</td>\n",
       "      <td>Progranulin</td>\n",
       "      <td>442-496</td>\n",
       "      <td>IGCDQHTSCPVGQTCCPSLGGSWACCQLPHAVCCEDRQHCCPAGYT...</td>\n",
       "      <td>Granulin-6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Secreted protein that acts as a key regulato...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>P01275</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21-50</td>\n",
       "      <td>RSLQDTEEKSRSFSASQADPLSDPDQMNED</td>\n",
       "      <td>Glicentin-related polypeptide</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Plays a key role in glucose metabolism and h...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    proteinID                chain interval  \\\n",
       "218    Q9UBU3                  NaN    76-98   \n",
       "219    P01009  Alpha-1-antitrypsin  375-418   \n",
       "220    P30990   Large neuromedin N  144-148   \n",
       "221    P30990   Large neuromedin N  166-170   \n",
       "223    P01282                  NaN   81-107   \n",
       "..        ...                  ...      ...   \n",
       "293    P28799          Progranulin  123-179   \n",
       "294    P28799          Progranulin  281-336   \n",
       "295    P28799          Progranulin  364-417   \n",
       "296    P28799          Progranulin  442-496   \n",
       "297    P01275                  NaN    21-50   \n",
       "\n",
       "                                               peptide  \\\n",
       "218                            FNAPFDVGIKLSGVQYQQHSQAL   \n",
       "219       MFLEAIPMSIPPEVKFNKPFVFLMIEQNTKSPLFMGKVVNPTQK   \n",
       "220                                              IPYIL   \n",
       "221                                              DSYYY   \n",
       "223                        HADGVFTSDFSKLLGQLSAKKYLESLM   \n",
       "..                                                 ...   \n",
       "293  AIQCPDSQFECPDFSTCCVMVDGSWGCCPMPQASCCEDRVHCCPHG...   \n",
       "294  DVKCDMEVSCPDGYTCCRLQSGAWGCCPFTQAVCCEDHIHCCPAGF...   \n",
       "295  VPCDNVSSCPSSDTCCQLTSGEWGCCPIPEAVCCSDHQHCCPQGYT...   \n",
       "296  IGCDQHTSCPVGQTCCPSLGGSWACCQLPHAVCCEDRQHCCPAGYT...   \n",
       "297                     RSLQDTEEKSRSFSASQADPLSDPDQMNED   \n",
       "\n",
       "                       description  \\\n",
       "218                      Obestatin   \n",
       "219         Short peptide from AAT   \n",
       "220                   Neuromedin N   \n",
       "221                   Tail peptide   \n",
       "223      Intestinal peptide PHM-27   \n",
       "..                             ...   \n",
       "293                     Granulin-2   \n",
       "294                     Granulin-4   \n",
       "295                     Granulin-5   \n",
       "296                     Granulin-6   \n",
       "297  Glicentin-related polypeptide   \n",
       "\n",
       "                                   associated_function  \\\n",
       "218  May be the ligand for GPR39. May have an appet...   \n",
       "219  Reversible chymotrypsin inhibitor. It also inh...   \n",
       "220                                                NaN   \n",
       "221                                                NaN   \n",
       "223                                                NaN   \n",
       "..                                                 ...   \n",
       "293                                                NaN   \n",
       "294  Promotes proliferation of the epithelial cell ...   \n",
       "295                                                NaN   \n",
       "296                                                NaN   \n",
       "297                                                NaN   \n",
       "\n",
       "                               non_associated_function  \n",
       "218                                                NaN  \n",
       "219                                                NaN  \n",
       "220  ['Neurotensin may play an endocrine or paracri...  \n",
       "221  ['Neurotensin may play an endocrine or paracri...  \n",
       "223  ['VIP causes vasodilation, lowers arterial blo...  \n",
       "..                                                 ...  \n",
       "293  ['Secreted protein that acts as a key regulato...  \n",
       "294                                                NaN  \n",
       "295  ['Secreted protein that acts as a key regulato...  \n",
       "296  ['Secreted protein that acts as a key regulato...  \n",
       "297  ['Plays a key role in glucose metabolism and h...  \n",
       "\n",
       "[63 rows x 7 columns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_reff_merged_final_me_df =no_reff_merged_final_df_qgrid.get_changed_df()\n",
    "no_reff_merged_final_me_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c3602c0",
   "metadata": {},
   "source": [
    "# Determines biological function using OpenAI/ChatGPT API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "1968158a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_info(response_string):\n",
    "    \"\"\"\n",
    "    Extracts function names, evidence, and logic from the provided response string.\n",
    "\n",
    "    Args:\n",
    "    - response_string (str): The response string containing function information.\n",
    "\n",
    "    Returns:\n",
    "    - list: A list of tuples where each tuple contains function name, logic, and evidence.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Split the response by the new line to separate the different functions\n",
    "    functions_list = response_string.split(\"\\n\\n\")\n",
    "    \n",
    "    # For each function, extract the function name, evidence, and logic\n",
    "    extracted_info = []\n",
    "    for func_str in functions_list:\n",
    "        if \"Function:\" in func_str:\n",
    "            func_name = func_str.split(\"Function:\")[1].split(\"|\")[0].strip()\n",
    "            evidence = func_str.split(\"Evidence:\")[1].split(\"|\")[0].strip()\n",
    "            logic = func_str.split(\"Logic:\")[1].split(\"|\")[0].strip()\n",
    "            extracted_info.append((func_name, logic, evidence))\n",
    "    \n",
    "    return extracted_info\n",
    "\n",
    "def classify_bioactivity(peptide, description, associated_function, non_associated_function, abstract, title, mbpdb_function_list):\n",
    "    \"\"\"\n",
    "    Constructs a query based on provided parameters, sends it to OpenAI's API, \n",
    "    and returns the extracted function information.\n",
    "\n",
    "    Args:\n",
    "    - peptide (str): The peptide of interest.\n",
    "    - description (str): Description of the peptide.\n",
    "    - associated_function (str): The associated function of the peptide.\n",
    "    - non_associated_function (str): Function of the protein or all peptides from that protein.\n",
    "    - abstract (str): The abstract related to the peptide.\n",
    "    - title (str): The title of the source.\n",
    "    - mbpdb_function_list (list): List of potential functions.\n",
    "\n",
    "    Returns:\n",
    "    - list: A list of tuples where each tuple contains function name, logic, and evidence.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Construct the query based on the provided parameters\n",
    "    query = (\n",
    "        f\"Given the peptide {peptide}, with the description: {description}, \"\n",
    "        f\"and the associated_function: {associated_function}, \"\n",
    "        f\"or function of the protein or all peptides from that protein called non_associated_function: {non_associated_function}, \"\n",
    "        f\"\\n\\nif associated_function = indeterminable then do ignore the non_associated_function input \"\n",
    "        f\"title: {title}, and abstract: {abstract}, \"\n",
    "        f\"please identify any bioactivity functions of the peptide from the list provided(mbpdb_function_list). \"\n",
    "        f\"For each identified function, please provide the response in the following format: \"\n",
    "        f\"'Function: [Function Name] | Evidence: [Specific evidence from the sources] | Logic: [Explanation for the choice]'\"\n",
    "        f\"\\n\\nList of potential functions:{mbpdb_function_list}.\"\n",
    "    )\n",
    "    \n",
    "    # Send the constructed query to OpenAI's API\n",
    "    response = openai.ChatCompletion.create(\n",
    "        model=\"gpt-4\",\n",
    "        messages=[\n",
    "            {\"role\": \"system\", \"content\": \"You are a helpful assistant.\"},\n",
    "            {\"role\": \"user\", \"content\": \"who is an expert in bioactive peptides\"},\n",
    "            {\"role\": \"user\", \"content\": query}\n",
    "        ]\n",
    "    )\n",
    "    \n",
    "    # Extract the response content\n",
    "    bioactivity_response = response['choices'][0]['message']['content']\n",
    "    \n",
    "    # Extract info from the response and return\n",
    "    return extract_info(bioactivity_response)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "d85df7fe",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed row 1/235 in 20.00 seconds\n",
      "Processed row 2/235 in 10.36 seconds\n",
      "Processed row 3/235 in 13.84 seconds\n",
      "Processed row 4/235 in 14.21 seconds\n",
      "Processed row 5/235 in 10.10 seconds\n",
      "Processed row 6/235 in 11.52 seconds\n",
      "Processed row 7/235 in 18.37 seconds\n",
      "Processed row 8/235 in 15.54 seconds\n",
      "Processed row 9/235 in 24.37 seconds\n",
      "Processed row 10/235 in 16.70 seconds\n",
      "Processed row 11/235 in 17.94 seconds\n",
      "Processed row 12/235 in 19.72 seconds\n",
      "Processed row 13/235 in 22.27 seconds\n",
      "Processed row 14/235 in 17.48 seconds\n",
      "Processed row 15/235 in 15.31 seconds\n",
      "Processed row 16/235 in 17.03 seconds\n",
      "Processed row 17/235 in 16.86 seconds\n",
      "Processed row 18/235 in 18.81 seconds\n",
      "Processed row 19/235 in 17.70 seconds\n",
      "Processed row 20/235 in 20.02 seconds\n",
      "Processed row 21/235 in 28.84 seconds\n",
      "Processed row 22/235 in 21.02 seconds\n",
      "Processed row 23/235 in 25.56 seconds\n",
      "Processed row 24/235 in 26.18 seconds\n",
      "Processed row 25/235 in 23.75 seconds\n",
      "Processed row 26/235 in 15.86 seconds\n",
      "Processed row 27/235 in 13.75 seconds\n",
      "Processed row 28/235 in 8.82 seconds\n",
      "Processed row 29/235 in 17.39 seconds\n",
      "Processed row 30/235 in 19.51 seconds\n",
      "Processed row 31/235 in 13.59 seconds\n",
      "Processed row 32/235 in 14.90 seconds\n",
      "Processed row 33/235 in 17.18 seconds\n",
      "Processed row 34/235 in 14.80 seconds\n",
      "Processed row 35/235 in 16.45 seconds\n",
      "Processed row 36/235 in 23.38 seconds\n",
      "Processed row 37/235 in 19.03 seconds\n",
      "Processed row 38/235 in 15.99 seconds\n",
      "Processed row 39/235 in 20.18 seconds\n",
      "Processed row 40/235 in 19.38 seconds\n",
      "Processed row 41/235 in 9.52 seconds\n",
      "Processed row 42/235 in 17.36 seconds\n",
      "Processed row 43/235 in 10.21 seconds\n",
      "Processed row 44/235 in 12.66 seconds\n",
      "Processed row 45/235 in 14.08 seconds\n",
      "Processed row 46/235 in 20.98 seconds\n",
      "Processed row 47/235 in 14.48 seconds\n",
      "Processed row 48/235 in 23.30 seconds\n",
      "Processed row 49/235 in 18.24 seconds\n",
      "Processed row 50/235 in 17.82 seconds\n",
      "Processed row 51/235 in 19.81 seconds\n",
      "Processed row 52/235 in 18.84 seconds\n",
      "Processed row 53/235 in 17.06 seconds\n",
      "Processed row 54/235 in 15.87 seconds\n",
      "Processed row 55/235 in 14.49 seconds\n",
      "Processed row 56/235 in 17.33 seconds\n",
      "Processed row 57/235 in 33.94 seconds\n",
      "Processed row 58/235 in 30.23 seconds\n",
      "Processed row 59/235 in 23.65 seconds\n",
      "Processed row 60/235 in 20.17 seconds\n",
      "Processed row 61/235 in 19.34 seconds\n",
      "Processed row 62/235 in 22.93 seconds\n",
      "Processed row 63/235 in 23.07 seconds\n",
      "Processed row 64/235 in 21.32 seconds\n",
      "Processed row 65/235 in 22.93 seconds\n",
      "Processed row 66/235 in 25.49 seconds\n",
      "Processed row 67/235 in 26.80 seconds\n",
      "Processed row 68/235 in 30.36 seconds\n",
      "Processed row 69/235 in 27.39 seconds\n",
      "Processed row 70/235 in 29.57 seconds\n",
      "Processed row 71/235 in 14.86 seconds\n",
      "Processed row 72/235 in 19.77 seconds\n",
      "Processed row 73/235 in 25.99 seconds\n",
      "Processed row 74/235 in 22.55 seconds\n",
      "Processed row 75/235 in 32.44 seconds\n",
      "Processed row 76/235 in 30.52 seconds\n",
      "Processed row 77/235 in 18.67 seconds\n",
      "Processed row 78/235 in 20.27 seconds\n",
      "Processed row 79/235 in 20.15 seconds\n",
      "Processed row 80/235 in 13.01 seconds\n",
      "Processed row 81/235 in 16.85 seconds\n",
      "Processed row 82/235 in 17.36 seconds\n",
      "Processed row 83/235 in 22.38 seconds\n",
      "Processed row 84/235 in 17.92 seconds\n",
      "Processed row 85/235 in 21.70 seconds\n",
      "Processed row 86/235 in 32.11 seconds\n",
      "Processed row 87/235 in 24.42 seconds\n",
      "Processed row 88/235 in 15.09 seconds\n",
      "Processed row 89/235 in 19.49 seconds\n",
      "Processed row 90/235 in 17.95 seconds\n",
      "Processed row 91/235 in 11.04 seconds\n",
      "Processed row 92/235 in 21.77 seconds\n",
      "Processed row 93/235 in 22.24 seconds\n",
      "Processed row 94/235 in 15.89 seconds\n",
      "Processed row 95/235 in 21.46 seconds\n",
      "Processed row 96/235 in 18.72 seconds\n",
      "Processed row 97/235 in 13.04 seconds\n",
      "Processed row 98/235 in 16.56 seconds\n",
      "Processed row 99/235 in 14.67 seconds\n",
      "Processed row 100/235 in 17.21 seconds\n",
      "Processed row 101/235 in 26.00 seconds\n",
      "Processed row 102/235 in 21.51 seconds\n",
      "Processed row 103/235 in 21.10 seconds\n",
      "Processed row 104/235 in 17.22 seconds\n",
      "Processed row 105/235 in 20.10 seconds\n",
      "Processed row 106/235 in 20.52 seconds\n",
      "Processed row 107/235 in 17.28 seconds\n",
      "Processed row 108/235 in 20.24 seconds\n",
      "Processed row 109/235 in 23.30 seconds\n",
      "Processed row 110/235 in 15.69 seconds\n",
      "Processed row 111/235 in 20.48 seconds\n",
      "Processed row 112/235 in 18.70 seconds\n",
      "Processed row 113/235 in 21.25 seconds\n",
      "Processed row 114/235 in 29.12 seconds\n",
      "Processed row 115/235 in 19.53 seconds\n",
      "Processed row 116/235 in 18.18 seconds\n",
      "Processed row 117/235 in 22.94 seconds\n",
      "Processed row 118/235 in 14.00 seconds\n",
      "Processed row 119/235 in 19.16 seconds\n",
      "Processed row 120/235 in 22.33 seconds\n",
      "Processed row 121/235 in 16.20 seconds\n",
      "Processed row 122/235 in 25.29 seconds\n",
      "Processed row 123/235 in 20.30 seconds\n",
      "Processed row 124/235 in 15.84 seconds\n",
      "Processed row 125/235 in 18.97 seconds\n",
      "Processed row 126/235 in 20.57 seconds\n",
      "Processed row 127/235 in 22.01 seconds\n",
      "Processed row 128/235 in 19.88 seconds\n",
      "Processed row 129/235 in 19.47 seconds\n",
      "Processed row 130/235 in 16.91 seconds\n",
      "Processed row 131/235 in 23.04 seconds\n",
      "Processed row 132/235 in 17.04 seconds\n",
      "Processed row 133/235 in 12.95 seconds\n",
      "Processed row 134/235 in 29.31 seconds\n",
      "Processed row 135/235 in 12.77 seconds\n",
      "Processed row 136/235 in 19.61 seconds\n",
      "Processed row 137/235 in 10.39 seconds\n",
      "Processed row 138/235 in 9.42 seconds\n",
      "Processed row 139/235 in 15.27 seconds\n",
      "Processed row 140/235 in 15.63 seconds\n",
      "Processed row 141/235 in 20.12 seconds\n",
      "Processed row 142/235 in 12.63 seconds\n",
      "Processed row 143/235 in 4.66 seconds\n",
      "Processed row 144/235 in 21.62 seconds\n",
      "Processed row 145/235 in 35.28 seconds\n",
      "Processed row 146/235 in 24.58 seconds\n",
      "Processed row 147/235 in 33.11 seconds\n",
      "Processed row 148/235 in 25.58 seconds\n",
      "Processed row 149/235 in 28.87 seconds\n",
      "Processed row 150/235 in 19.21 seconds\n",
      "Processed row 151/235 in 31.18 seconds\n",
      "Processed row 152/235 in 15.23 seconds\n",
      "Processed row 153/235 in 29.03 seconds\n",
      "Processed row 154/235 in 21.06 seconds\n",
      "Processed row 155/235 in 22.36 seconds\n",
      "Processed row 156/235 in 22.55 seconds\n",
      "Processed row 157/235 in 26.28 seconds\n",
      "Processed row 158/235 in 21.29 seconds\n",
      "Processed row 159/235 in 20.51 seconds\n",
      "Processed row 160/235 in 27.76 seconds\n",
      "Processed row 161/235 in 21.42 seconds\n",
      "Processed row 162/235 in 13.81 seconds\n",
      "Processed row 163/235 in 19.49 seconds\n",
      "Processed row 164/235 in 22.12 seconds\n",
      "Processed row 165/235 in 26.82 seconds\n",
      "Processed row 166/235 in 20.05 seconds\n",
      "Processed row 167/235 in 26.87 seconds\n",
      "Processed row 168/235 in 23.23 seconds\n",
      "Processed row 169/235 in 26.74 seconds\n",
      "Processed row 170/235 in 23.25 seconds\n",
      "Processed row 171/235 in 12.66 seconds\n",
      "Processed row 172/235 in 23.24 seconds\n",
      "Processed row 173/235 in 24.45 seconds\n",
      "Processed row 174/235 in 14.61 seconds\n",
      "Processed row 175/235 in 23.76 seconds\n",
      "Processed row 176/235 in 17.64 seconds\n",
      "Processed row 177/235 in 17.19 seconds\n",
      "Processed row 178/235 in 21.36 seconds\n",
      "Processed row 179/235 in 25.85 seconds\n",
      "Processed row 180/235 in 19.19 seconds\n",
      "Processed row 181/235 in 23.71 seconds\n",
      "Processed row 182/235 in 18.36 seconds\n",
      "Processed row 183/235 in 32.82 seconds\n",
      "Processed row 184/235 in 35.46 seconds\n",
      "Processed row 185/235 in 26.56 seconds\n",
      "Processed row 186/235 in 20.94 seconds\n",
      "Processed row 187/235 in 13.75 seconds\n",
      "Processed row 188/235 in 16.19 seconds\n",
      "Processed row 189/235 in 13.36 seconds\n",
      "Processed row 190/235 in 7.22 seconds\n",
      "Processed row 191/235 in 11.95 seconds\n",
      "Processed row 192/235 in 13.87 seconds\n",
      "Processed row 193/235 in 9.98 seconds\n",
      "Processed row 194/235 in 20.08 seconds\n",
      "Processed row 195/235 in 12.93 seconds\n",
      "Processed row 196/235 in 12.11 seconds\n",
      "Processed row 197/235 in 13.58 seconds\n",
      "Processed row 198/235 in 20.62 seconds\n",
      "Processed row 199/235 in 19.01 seconds\n",
      "Processed row 200/235 in 15.83 seconds\n",
      "Processed row 201/235 in 19.05 seconds\n",
      "Processed row 202/235 in 11.85 seconds\n",
      "Processed row 203/235 in 12.21 seconds\n",
      "Processed row 204/235 in 17.22 seconds\n",
      "Processed row 205/235 in 18.24 seconds\n",
      "Processed row 206/235 in 28.85 seconds\n",
      "Processed row 207/235 in 27.49 seconds\n",
      "Processed row 208/235 in 20.61 seconds\n",
      "Processed row 209/235 in 20.61 seconds\n",
      "Processed row 210/235 in 21.90 seconds\n",
      "Processed row 211/235 in 32.31 seconds\n",
      "Processed row 212/235 in 22.68 seconds\n",
      "Processed row 213/235 in 24.04 seconds\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed row 214/235 in 18.77 seconds\n",
      "Processed row 215/235 in 24.89 seconds\n",
      "Processed row 216/235 in 14.64 seconds\n",
      "Processed row 217/235 in 10.16 seconds\n",
      "Processed row 218/235 in 13.23 seconds\n",
      "Processed row 219/235 in 16.46 seconds\n",
      "Processed row 220/235 in 26.62 seconds\n",
      "Processed row 221/235 in 13.50 seconds\n",
      "Processed row 222/235 in 13.78 seconds\n",
      "Processed row 223/235 in 18.29 seconds\n",
      "Processed row 224/235 in 12.18 seconds\n",
      "Processed row 225/235 in 10.13 seconds\n",
      "Processed row 226/235 in 11.68 seconds\n",
      "Processed row 227/235 in 14.89 seconds\n",
      "Processed row 228/235 in 11.04 seconds\n",
      "Processed row 229/235 in 11.40 seconds\n",
      "Processed row 230/235 in 10.58 seconds\n",
      "Processed row 231/235 in 17.26 seconds\n",
      "Processed row 232/235 in 14.82 seconds\n",
      "Processed row 233/235 in 11.26 seconds\n",
      "Processed row 234/235 in 11.00 seconds\n",
      "Processed row 235/235 in 11.95 seconds\n"
     ]
    }
   ],
   "source": [
    "# Create an empty list to store results\n",
    "results = []\n",
    "\n",
    "# Iterate over each row of the DataFrame\n",
    "# Using enumerate to get both the row number (index) and the row data\n",
    "for index, (_, row) in enumerate(reff_merged_final_me_df.iterrows()):\n",
    "    # Record the start time for this iteration\n",
    "    start_time = time.time()\n",
    "    \n",
    "    # Get the functions, logic, and evidence based on the row data\n",
    "    functions_with_logic = classify_bioactivity(row['peptide'], row['description'], row['associated_function'], row['non_associated_function'], row['abstract'], row['title'], mbpdb_function_list)\n",
    "    \n",
    "    # For each function, logic, and evidence, create a new row and append it to the results list\n",
    "    for func, logic, evidence in functions_with_logic:\n",
    "        new_row = row.copy()\n",
    "        new_row['gpt_function'] = func\n",
    "        new_row['gpt_logic'] = logic\n",
    "        new_row['gpt_evidence'] = evidence\n",
    "        results.append(new_row)\n",
    "    \n",
    "    # Calculate the elapsed time for this iteration\n",
    "    elapsed_time = time.time() - start_time\n",
    "    print(f\"Processed row {index + 1}/{len(reff_merged_final_me_df)} in {elapsed_time:.2f} seconds\")\n",
    "\n",
    "# Convert the results to a new DataFrame\n",
    "reff_merged_final_me_gpt_df = pd.DataFrame(results)\n",
    "\n",
    "# Display the updated DataFrame using qgrid\n",
    "# updated_df = qgrid.show_grid(updated_df, show_toolbar=True, grid_options={'forceFitColumns': False})\n",
    "# updated_df        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "3e925204",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e3f4f7c8ad7f46ca93cae69c05d17e8a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': False, 'defa…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Display the updated DataFrame using qgrid\n",
    "reff_merged_final_me_gpt_df_qgrid = qgrid.show_grid(reff_merged_final_me_gpt_df, show_toolbar=True, grid_options={'forceFitColumns': False})\n",
    "reff_merged_final_me_gpt_df_qgrid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "49119760",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Pain perception',\n",
       " 'ACE-Inhibitory',\n",
       " 'Antithrombotic',\n",
       " 'Improve cognition',\n",
       " 'Immunomodulatory',\n",
       " 'Hormone secreted by pancreatic cells that acts as a regulator of pancreatic and gastrointestinal functions probably by signaling through the G protein-coupled receptor NPY4R2.',\n",
       " 'Diuretic',\n",
       " 'Vasoconstrictory action',\n",
       " 'Cell Penetrating',\n",
       " 'Inhibition of forskolin-stimulated cAMP synthesis.',\n",
       " 'Potent stimulator of insulin secretion',\n",
       " 'Ameliorates Insulin Resistance',\n",
       " 'Increase exocrine pancreatic secretion',\n",
       " 'Increase mucin secretion',\n",
       " 'DNA repair',\n",
       " 'Enhanced insulin signaling',\n",
       " 'Antimicrobial',\n",
       " 'Antioxidant',\n",
       " 'Nitric oxide liberation',\n",
       " 'Ameliorates insulin resistance',\n",
       " 'Enhance Insulin Signaling',\n",
       " 'Increase calcium uptake',\n",
       " 'Satiety',\n",
       " 'Opioid',\n",
       " 'Cytomodulatory',\n",
       " 'Improves cognition',\n",
       " 'Increase cellular growth',\n",
       " 'Antihypertensive',\n",
       " 'Prolyl endopeptidase-inhibitory',\n",
       " 'ACE-inhibitory',\n",
       " 'Increase intestinal motility',\n",
       " 'Cytotoxic',\n",
       " 'Bradykinin-Potentiating',\n",
       " 'Wound Healing',\n",
       " 'CytoModulatory',\n",
       " 'Enhance insulin signaling',\n",
       " 'Ameliorates insulin resistance, Satiety, Antianxiety',\n",
       " 'Wound healing',\n",
       " 'Decrease intestinal motility',\n",
       " 'Antiapoptotic effect',\n",
       " 'Antianxiety',\n",
       " 'Anticancer',\n",
       " 'Stress Response',\n",
       " 'Contraction of smooth muscle of uterus and mammary gland',\n",
       " 'Cholesterol regulation']"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Defining the cleaning function\n",
    "def clean_gpt_function(func):\n",
    "    return func.split(\"\\n\")[0].strip().strip(\"'\")\n",
    "\n",
    "# Assuming reff_merged_final_me_gpt_df is already defined, applying the cleaning function to its 'gpt_function' column\n",
    "reff_merged_final_me_gpt_df['gpt_function'] = reff_merged_final_me_gpt_df['gpt_function'].apply(clean_gpt_function)\n",
    "list(set(reff_merged_final_me_gpt_df['gpt_function']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7a4eb1bd",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'rr' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_16724/1101791251.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mrr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'reff_merged_final_me_gpt_df.csv'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[0mprint_critical_info\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mreff_merged_final_me_gpt_df\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'reff_merged_final_me_gpt_df'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'rr' is not defined"
     ]
    }
   ],
   "source": [
    "reff_merged_final_me_gpt_df.to_csv('data/reff_merged_final_me_gpt_df.csv')\n",
    "print_critical_info(reff_merged_final_me_gpt_df,'data/reff_merged_final_me_gpt_df')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8cf6dddc",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'reff_merged_final_me_gpt_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_16724/2965572109.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;31m#updated_df = updated_df.get_changed_df()\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;31m#updated_df.to_csv('uni_reff_merged_final_me_gpt_df.csv')\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 4\u001b[1;33m \u001b[0mupdated_df\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mreff_merged_final_me_gpt_df\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'reff_merged_final_me_gpt_df' is not defined"
     ]
    }
   ],
   "source": [
    "#updated_df = qgrid.show_grid(updated_df, show_toolbar=True, grid_options={'forceFitColumns': False})\n",
    "#updated_df = updated_df.get_changed_df()\n",
    "#updated_df.to_csv('data/uni_reff_merged_final_me_gpt_df.csv')\n",
    "updated_df = reff_merged_final_me_gpt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "0dab7418",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Information for DataFrame: b4merge\n",
      "----------------------------------------\n",
      "Numbers of peptides: 22\n",
      "Shape: (22, 15)\n",
      "Columns: ['proteinID', 'chain', 'interval', 'peptide', 'description', 'associated_function', 'non_associated_function', 'title', 'authors', 'doi', 'pubmed_id', 'abstract', 'function', 'logic', 'evidence']\n",
      "\n",
      "Missing Values Count:\n",
      "proteinID                   0\n",
      "chain                       0\n",
      "interval                    0\n",
      "peptide                     0\n",
      "description                 0\n",
      "associated_function         0\n",
      "non_associated_function    20\n",
      "title                       0\n",
      "authors                     0\n",
      "doi                         2\n",
      "pubmed_id                   2\n",
      "abstract                    2\n",
      "function                    0\n",
      "logic                       0\n",
      "evidence                    0\n",
      "dtype: int64\n",
      "\n",
      "========================================\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#reff_merged_final_me_gpt_df=pd.read_csv('reff_merged_final_me_gpt_df.csv', index_col=0)\n",
    "print_critical_info(updated_df,\"b4merge\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "fb7481b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18, 12)\n",
      "(18, 12)\n"
     ]
    }
   ],
   "source": [
    "print(processed_peptide_df.shape)\n",
    "processed_peptide_df = processed_peptide_df.drop_duplicates(subset=['proteinID', 'peptide'])\n",
    "print(processed_peptide_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "f6e6dc2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Information for DataFrame: aftermerge\n",
      "----------------------------------------\n",
      "Numbers of peptides: 22\n",
      "Shape: (22, 17)\n",
      "Columns: ['proteinID', 'protein_name', 'protein_desc', 'chain', 'interval', 'peptide', 'description', 'associated_function', 'non_associated_function', 'title', 'authors', 'doi', 'pubmed_id', 'abstract', 'function', 'logic', 'evidence']\n",
      "\n",
      "Missing Values Count:\n",
      "proteinID                   0\n",
      "protein_name                0\n",
      "protein_desc                0\n",
      "chain                       0\n",
      "interval                    0\n",
      "peptide                     0\n",
      "description                 0\n",
      "associated_function         0\n",
      "non_associated_function    20\n",
      "title                       0\n",
      "authors                     0\n",
      "doi                         2\n",
      "pubmed_id                   2\n",
      "abstract                    2\n",
      "function                    0\n",
      "logic                       0\n",
      "evidence                    0\n",
      "dtype: int64\n",
      "\n",
      "========================================\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Merge based on 'proteinID' and 'peptide' using an inner merge\n",
    "merged_df = updated_df.merge(processed_peptide_df[['proteinID', 'peptide', 'protein_name', 'protein_desc']], \n",
    "                             on=['proteinID', 'peptide'], \n",
    "                             how='inner')\n",
    "\n",
    "# Rearrange columns\n",
    "cols = list(updated_df.columns) + ['protein_name', 'protein_desc']\n",
    "updated_df = merged_df[cols]\n",
    "\n",
    "# Extract current columns\n",
    "cols = list(updated_df.columns)\n",
    "\n",
    "# Reorder columns\n",
    "cols = [cols[0]] + ['protein_name', 'protein_desc'] + cols[1:-2]\n",
    "\n",
    "# Assign the reordered columns to updated_df\n",
    "updated_df = updated_df[cols]\n",
    "print_critical_info(updated_df,\"aftermerge\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "4ca30c8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>protein_desc</th>\n",
       "      <th>chain</th>\n",
       "      <th>interval</th>\n",
       "      <th>peptide</th>\n",
       "      <th>description</th>\n",
       "      <th>associated_function</th>\n",
       "      <th>non_associated_function</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>doi</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>abstract</th>\n",
       "      <th>function</th>\n",
       "      <th>logic</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P02788</td>\n",
       "      <td>TRFL_HUMAN</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>171-201</td>\n",
       "      <td>FFSASCVPGADKGQFPNLCRLCAGTGENKCA</td>\n",
       "      <td>Kaliocin-1</td>\n",
       "      <td>Has antimicrobial activity and is able to perm...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Potassium efflux induced by a new lactoferrin-...</td>\n",
       "      <td>Viejo-Diaz, M. et al.</td>\n",
       "      <td>10.1023/a:1022657630698</td>\n",
       "      <td>12693969.0</td>\n",
       "      <td>A 31-amino acid synthetic peptide (NH(2)-FFSAS...</td>\n",
       "      <td>Cell Penetrating</td>\n",
       "      <td>The ability of the peptide to induce the efflu...</td>\n",
       "      <td>The antimicrobial effect of kaliocin-1 was low...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P02788</td>\n",
       "      <td>TRFL_HUMAN</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>171-201</td>\n",
       "      <td>FFSASCVPGADKGQFPNLCRLCAGTGENKCA</td>\n",
       "      <td>Kaliocin-1</td>\n",
       "      <td>Has antimicrobial activity and is able to perm...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Potassium efflux induced by a new lactoferrin-...</td>\n",
       "      <td>Viejo-Diaz, M. et al.</td>\n",
       "      <td>10.1023/a:1022657630698</td>\n",
       "      <td>12693969.0</td>\n",
       "      <td>A 31-amino acid synthetic peptide (NH(2)-FFSAS...</td>\n",
       "      <td>Antimicrobial</td>\n",
       "      <td>The antimicrobial property is evident from the...</td>\n",
       "      <td>The peptide, termed kaliocin-1, and lactoferri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>41-45</td>\n",
       "      <td>INKKI</td>\n",
       "      <td>Casoparan</td>\n",
       "      <td>Casoparan acts as a macrophage activator, incr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Effects of 'casoparan', a peptide isolated fro...</td>\n",
       "      <td>Lebrun, I. et al.</td>\n",
       "      <td>10.1080/09629350400003068</td>\n",
       "      <td>15545057.0</td>\n",
       "      <td>Casein, a protein found in milk of several spe...</td>\n",
       "      <td>'Bradykinin-Potentiating'</td>\n",
       "      <td>The peptide's ability to enhance the actions o...</td>\n",
       "      <td>\"It also acts as a bradykinin-potentiating pep...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>41-45</td>\n",
       "      <td>INKKI</td>\n",
       "      <td>Casoparan</td>\n",
       "      <td>Casoparan acts as a macrophage activator, incr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Effects of 'casoparan', a peptide isolated fro...</td>\n",
       "      <td>Lebrun, I. et al.</td>\n",
       "      <td>10.1080/09629350400003068</td>\n",
       "      <td>15545057.0</td>\n",
       "      <td>Casein, a protein found in milk of several spe...</td>\n",
       "      <td>'Immunomodulatory'</td>\n",
       "      <td>The peptide's ability to increase the activity...</td>\n",
       "      <td>\"Casoparan acts as a macrophage activator, inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>41-45</td>\n",
       "      <td>INKKI</td>\n",
       "      <td>Casoparan</td>\n",
       "      <td>Casoparan acts as a macrophage activator, incr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Effects of 'casoparan', a peptide isolated fro...</td>\n",
       "      <td>Lebrun, I. et al.</td>\n",
       "      <td>10.1080/09629350400003068</td>\n",
       "      <td>15545057.0</td>\n",
       "      <td>Casein, a protein found in milk of several spe...</td>\n",
       "      <td>'Cytomodulatory'</td>\n",
       "      <td>The peptide's direct influence on the activity...</td>\n",
       "      <td>\"The immunostimulating action is probably rela...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P02663</td>\n",
       "      <td>CASA2_BOVIN</td>\n",
       "      <td>Alpha-S2-casein</td>\n",
       "      <td>Alpha-S2-casein</td>\n",
       "      <td>165-203</td>\n",
       "      <td>KTKLTEEEKNRLNFLKKISQRYQKFALPQYLKTVYQHQK</td>\n",
       "      <td>Casocidin-1</td>\n",
       "      <td>Casocidin-I inhibits the growth of E.coli and ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Casocidin-I: a casein-alpha s2 derived peptide...</td>\n",
       "      <td>Zucht, H. et al.</td>\n",
       "      <td>10.1016/0014-5793(95)00974-e</td>\n",
       "      <td>7556666.0</td>\n",
       "      <td>Here we report the isolation and characterizat...</td>\n",
       "      <td>Antimicrobial</td>\n",
       "      <td>The evidence for the antimicrobial function co...</td>\n",
       "      <td>The abstract reports the isolation and charact...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>129-136</td>\n",
       "      <td>YPVEPFTE</td>\n",
       "      <td>Casohypotensin</td>\n",
       "      <td>Casohypotensin acts as a bradykinin-potentiati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biochemical and pharmacological aspects of two...</td>\n",
       "      <td>Perpetuo, E. et al.</td>\n",
       "      <td>10.1023/b:jopc.0000008724.98339.ff</td>\n",
       "      <td>14714726.0</td>\n",
       "      <td>Peptides that display bradykinin-potentiating ...</td>\n",
       "      <td>ACE-inhibitory</td>\n",
       "      <td>This means that the peptide is competitive, wh...</td>\n",
       "      <td>The octapeptide YPVEPFTE showed an in vitro co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>129-136</td>\n",
       "      <td>YPVEPFTE</td>\n",
       "      <td>Casohypotensin</td>\n",
       "      <td>Casohypotensin acts as a bradykinin-potentiati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biochemical and pharmacological aspects of two...</td>\n",
       "      <td>Perpetuo, E. et al.</td>\n",
       "      <td>10.1023/b:jopc.0000008724.98339.ff</td>\n",
       "      <td>14714726.0</td>\n",
       "      <td>Peptides that display bradykinin-potentiating ...</td>\n",
       "      <td>Bradykinin-Potentiating</td>\n",
       "      <td>The fact that the peptide potentiates the acti...</td>\n",
       "      <td>The peptide YPVEPFTE has been described as a b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>129-136</td>\n",
       "      <td>YPVEPFTE</td>\n",
       "      <td>Casohypotensin</td>\n",
       "      <td>Casohypotensin acts as a bradykinin-potentiati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biochemical and pharmacological aspects of two...</td>\n",
       "      <td>Perpetuo, E. et al.</td>\n",
       "      <td>10.1023/b:jopc.0000008724.98339.ff</td>\n",
       "      <td>14714726.0</td>\n",
       "      <td>Peptides that display bradykinin-potentiating ...</td>\n",
       "      <td>Opioid</td>\n",
       "      <td>The observed effect implies that the peptide f...</td>\n",
       "      <td>The same peptide presented opiate-like activit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>129-136</td>\n",
       "      <td>YPVEPFTE</td>\n",
       "      <td>Casohypotensin</td>\n",
       "      <td>Casohypotensin acts as a bradykinin-potentiati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biochemical and pharmacological aspects of two...</td>\n",
       "      <td>Perpetuo, E. et al.</td>\n",
       "      <td>10.1023/b:jopc.0000008724.98339.ff</td>\n",
       "      <td>14714726.0</td>\n",
       "      <td>Peptides that display bradykinin-potentiating ...</td>\n",
       "      <td>Antihypertensive</td>\n",
       "      <td>This function results from the capacity of bra...</td>\n",
       "      <td>Casohypotensin acts as a bradykinin-potentiati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>129-136</td>\n",
       "      <td>YPVEPFTE</td>\n",
       "      <td>Casohypotensin</td>\n",
       "      <td>Casohypotensin acts as a bradykinin-potentiati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolation and characterization of a new bradyk...</td>\n",
       "      <td>Lebrun, I. et al.</td>\n",
       "      <td>10.1139/y95-012</td>\n",
       "      <td>7600458.0</td>\n",
       "      <td>Peptides that display bradykinin-potentiating ...</td>\n",
       "      <td>Antihypertensive</td>\n",
       "      <td>Hypotension, or reduced blood pressure, is ind...</td>\n",
       "      <td>Intravenous injections of bradykinin and of th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>129-136</td>\n",
       "      <td>YPVEPFTE</td>\n",
       "      <td>Casohypotensin</td>\n",
       "      <td>Casohypotensin acts as a bradykinin-potentiati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolation and characterization of a new bradyk...</td>\n",
       "      <td>Lebrun, I. et al.</td>\n",
       "      <td>10.1139/y95-012</td>\n",
       "      <td>7600458.0</td>\n",
       "      <td>Peptides that display bradykinin-potentiating ...</td>\n",
       "      <td>Bradykinin-Potentiating</td>\n",
       "      <td>The peptide's interaction with bradykinin, an ...</td>\n",
       "      <td>The octapeptide Tyr-Pro-Val-Gln-Pro-Phe-Thr-Gl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>129-136</td>\n",
       "      <td>YPVEPFTE</td>\n",
       "      <td>Casohypotensin</td>\n",
       "      <td>Casohypotensin acts as a bradykinin-potentiati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolation and characterization of a new bradyk...</td>\n",
       "      <td>Lebrun, I. et al.</td>\n",
       "      <td>10.1139/y95-012</td>\n",
       "      <td>7600458.0</td>\n",
       "      <td>Peptides that display bradykinin-potentiating ...</td>\n",
       "      <td>Prolyl endopeptidase-inhibitory</td>\n",
       "      <td>The peptide's ability to inhibit endo-oligopep...</td>\n",
       "      <td>This bradykinin-potentiating octapeptide is a ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>P02788</td>\n",
       "      <td>TRFL_HUMAN</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>338-343</td>\n",
       "      <td>YLGSGY</td>\n",
       "      <td>Lactoferroxin-A</td>\n",
       "      <td>Has opioid antagonist activity (PubMed:1369293...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolation and characterization of opioid antag...</td>\n",
       "      <td>Tani, F. et al.</td>\n",
       "      <td>10.1271/bbb1961.54.1803</td>\n",
       "      <td>1369293.0</td>\n",
       "      <td>Peptides with affinity for opioid receptors we...</td>\n",
       "      <td>Opioid</td>\n",
       "      <td>The term opioid refers to compounds with narco...</td>\n",
       "      <td>The peptide YLGSGY, also known as Lactoferroxi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>P02788</td>\n",
       "      <td>TRFL_HUMAN</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>543-547</td>\n",
       "      <td>RYYGY</td>\n",
       "      <td>Lactoferroxin-B</td>\n",
       "      <td>Has opioid antagonist activity (PubMed:1369293...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolation and characterization of opioid antag...</td>\n",
       "      <td>Tani, F. et al.</td>\n",
       "      <td>10.1271/bbb1961.54.1803</td>\n",
       "      <td>1369293.0</td>\n",
       "      <td>Peptides with affinity for opioid receptors we...</td>\n",
       "      <td>Opioid</td>\n",
       "      <td>Based on the evidence provided, this peptide d...</td>\n",
       "      <td>The peptide RYYGY (Lactoferroxin-B) has affini...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>P02788</td>\n",
       "      <td>TRFL_HUMAN</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>680-686</td>\n",
       "      <td>KYLGPQY</td>\n",
       "      <td>Lactoferroxin-C</td>\n",
       "      <td>Has opioid antagonist activity (PubMed:1369293...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolation and characterization of opioid antag...</td>\n",
       "      <td>Tani, F. et al.</td>\n",
       "      <td>10.1271/bbb1961.54.1803</td>\n",
       "      <td>1369293.0</td>\n",
       "      <td>Peptides with affinity for opioid receptors we...</td>\n",
       "      <td>Opioid</td>\n",
       "      <td>The peptide is directly described as having op...</td>\n",
       "      <td>The peptide has been found to have an affinity...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>P02668</td>\n",
       "      <td>CASK_BOVIN</td>\n",
       "      <td>Kappa-casein</td>\n",
       "      <td>Kappa-casein</td>\n",
       "      <td>54-59</td>\n",
       "      <td>SRYPSY</td>\n",
       "      <td>Casoxin-6</td>\n",
       "      <td>indeterminable</td>\n",
       "      <td>['Kappa-casein stabilizes micelle formation, p...</td>\n",
       "      <td>Specificity of hydrolysis of bovine kappa-case...</td>\n",
       "      <td>Reid, J. et al.</td>\n",
       "      <td>10.1128/aem.60.3.801-806.1994</td>\n",
       "      <td>8161175.0</td>\n",
       "      <td>The cell envelope-associated proteinases from ...</td>\n",
       "      <td>Antihypertensive</td>\n",
       "      <td>As SRYPSY comes from Kappa-casein and given th...</td>\n",
       "      <td>The action of the cell envelope-associated pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>P02668</td>\n",
       "      <td>CASK_BOVIN</td>\n",
       "      <td>Kappa-casein</td>\n",
       "      <td>Kappa-casein</td>\n",
       "      <td>54-59</td>\n",
       "      <td>SRYPSY</td>\n",
       "      <td>Casoxin-6</td>\n",
       "      <td>indeterminable</td>\n",
       "      <td>['Kappa-casein stabilizes micelle formation, p...</td>\n",
       "      <td>Specificity of hydrolysis of bovine kappa-case...</td>\n",
       "      <td>Reid, J. et al.</td>\n",
       "      <td>10.1128/aem.60.3.801-806.1994</td>\n",
       "      <td>8161175.0</td>\n",
       "      <td>The cell envelope-associated proteinases from ...</td>\n",
       "      <td>Opioid</td>\n",
       "      <td>As SRYPSY is a casoxin-6 peptide, it might pos...</td>\n",
       "      <td>Casoxins A, B, and C have opioid antagonist ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>P24627</td>\n",
       "      <td>TRFL_BOVIN</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>36-60</td>\n",
       "      <td>FKCRRWQWRMKKLGAPSITCVRRAF</td>\n",
       "      <td>Lactoferricin-B</td>\n",
       "      <td>Lactoferricin B is an antimicrobial peptide. I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Identification of the bactericidal domain of l...</td>\n",
       "      <td>Bellamy, W. et al.</td>\n",
       "      <td>10.1016/0167-4838(92)90346-f</td>\n",
       "      <td>1599934.0</td>\n",
       "      <td>We report the existence of a previously unknow...</td>\n",
       "      <td>Antimicrobial</td>\n",
       "      <td>The peptide has been proven to inhibit the gro...</td>\n",
       "      <td>The description provided states \"Lactoferricin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>P24627</td>\n",
       "      <td>TRFL_BOVIN</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>36-60</td>\n",
       "      <td>FKCRRWQWRMKKLGAPSITCVRRAF</td>\n",
       "      <td>Lactoferricin-B</td>\n",
       "      <td>Lactoferricin B is an antimicrobial peptide. I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Identification of the bactericidal domain of l...</td>\n",
       "      <td>Bellamy, W. et al.</td>\n",
       "      <td>10.1016/0167-4838(92)90346-f</td>\n",
       "      <td>1599934.0</td>\n",
       "      <td>We report the existence of a previously unknow...</td>\n",
       "      <td>Cytotoxic</td>\n",
       "      <td>The peptide causes a rapid loss of colony-form...</td>\n",
       "      <td>\"The effect of the isolated domain was lethal ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>P02662</td>\n",
       "      <td>CASA1_BOVIN</td>\n",
       "      <td>Alpha-S1-casein</td>\n",
       "      <td>Alpha-S1-casein</td>\n",
       "      <td>95-105</td>\n",
       "      <td>HIQKEDVPSER</td>\n",
       "      <td>Antioxidant peptide</td>\n",
       "      <td>Antioxidant peptide has 2,2-diphenyl-1-picrylh...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Antioxidant peptides isolated from cheddar che...</td>\n",
       "      <td>Gupta, A. et al.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Antioxidant</td>\n",
       "      <td>Given that the associated function is stated t...</td>\n",
       "      <td>The peptide HIQKEDVPSER is described as an ant...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>113-120</td>\n",
       "      <td>VKEAMAPK</td>\n",
       "      <td>Antioxidant peptide</td>\n",
       "      <td>Antioxidant peptide has antioxidant activity.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Studies on antioxidative peptides generated in...</td>\n",
       "      <td>Gupta, A. et al.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Antioxidant</td>\n",
       "      <td>The peptide is explicitly described as an anti...</td>\n",
       "      <td>The given peptide VKEAMAPK is described as an ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   proteinID protein_name      protein_desc             chain interval  \\\n",
       "0     P02788   TRFL_HUMAN  Lactotransferrin  Lactotransferrin  171-201   \n",
       "1     P02788   TRFL_HUMAN  Lactotransferrin  Lactotransferrin  171-201   \n",
       "2     P02666   CASB_BOVIN       Beta-casein       Beta-casein    41-45   \n",
       "3     P02666   CASB_BOVIN       Beta-casein       Beta-casein    41-45   \n",
       "4     P02666   CASB_BOVIN       Beta-casein       Beta-casein    41-45   \n",
       "5     P02663  CASA2_BOVIN   Alpha-S2-casein   Alpha-S2-casein  165-203   \n",
       "6     P02666   CASB_BOVIN       Beta-casein       Beta-casein  129-136   \n",
       "7     P02666   CASB_BOVIN       Beta-casein       Beta-casein  129-136   \n",
       "8     P02666   CASB_BOVIN       Beta-casein       Beta-casein  129-136   \n",
       "9     P02666   CASB_BOVIN       Beta-casein       Beta-casein  129-136   \n",
       "10    P02666   CASB_BOVIN       Beta-casein       Beta-casein  129-136   \n",
       "11    P02666   CASB_BOVIN       Beta-casein       Beta-casein  129-136   \n",
       "12    P02666   CASB_BOVIN       Beta-casein       Beta-casein  129-136   \n",
       "13    P02788   TRFL_HUMAN  Lactotransferrin  Lactotransferrin  338-343   \n",
       "14    P02788   TRFL_HUMAN  Lactotransferrin  Lactotransferrin  543-547   \n",
       "15    P02788   TRFL_HUMAN  Lactotransferrin  Lactotransferrin  680-686   \n",
       "16    P02668   CASK_BOVIN      Kappa-casein      Kappa-casein    54-59   \n",
       "17    P02668   CASK_BOVIN      Kappa-casein      Kappa-casein    54-59   \n",
       "18    P24627   TRFL_BOVIN  Lactotransferrin  Lactotransferrin    36-60   \n",
       "19    P24627   TRFL_BOVIN  Lactotransferrin  Lactotransferrin    36-60   \n",
       "20    P02662  CASA1_BOVIN   Alpha-S1-casein   Alpha-S1-casein   95-105   \n",
       "21    P02666   CASB_BOVIN       Beta-casein       Beta-casein  113-120   \n",
       "\n",
       "                                    peptide          description  \\\n",
       "0           FFSASCVPGADKGQFPNLCRLCAGTGENKCA           Kaliocin-1   \n",
       "1           FFSASCVPGADKGQFPNLCRLCAGTGENKCA           Kaliocin-1   \n",
       "2                                     INKKI            Casoparan   \n",
       "3                                     INKKI            Casoparan   \n",
       "4                                     INKKI            Casoparan   \n",
       "5   KTKLTEEEKNRLNFLKKISQRYQKFALPQYLKTVYQHQK          Casocidin-1   \n",
       "6                                  YPVEPFTE       Casohypotensin   \n",
       "7                                  YPVEPFTE       Casohypotensin   \n",
       "8                                  YPVEPFTE       Casohypotensin   \n",
       "9                                  YPVEPFTE       Casohypotensin   \n",
       "10                                 YPVEPFTE       Casohypotensin   \n",
       "11                                 YPVEPFTE       Casohypotensin   \n",
       "12                                 YPVEPFTE       Casohypotensin   \n",
       "13                                   YLGSGY      Lactoferroxin-A   \n",
       "14                                    RYYGY      Lactoferroxin-B   \n",
       "15                                  KYLGPQY      Lactoferroxin-C   \n",
       "16                                   SRYPSY            Casoxin-6   \n",
       "17                                   SRYPSY            Casoxin-6   \n",
       "18                FKCRRWQWRMKKLGAPSITCVRRAF      Lactoferricin-B   \n",
       "19                FKCRRWQWRMKKLGAPSITCVRRAF      Lactoferricin-B   \n",
       "20                              HIQKEDVPSER  Antioxidant peptide   \n",
       "21                                 VKEAMAPK  Antioxidant peptide   \n",
       "\n",
       "                                  associated_function  \\\n",
       "0   Has antimicrobial activity and is able to perm...   \n",
       "1   Has antimicrobial activity and is able to perm...   \n",
       "2   Casoparan acts as a macrophage activator, incr...   \n",
       "3   Casoparan acts as a macrophage activator, incr...   \n",
       "4   Casoparan acts as a macrophage activator, incr...   \n",
       "5   Casocidin-I inhibits the growth of E.coli and ...   \n",
       "6   Casohypotensin acts as a bradykinin-potentiati...   \n",
       "7   Casohypotensin acts as a bradykinin-potentiati...   \n",
       "8   Casohypotensin acts as a bradykinin-potentiati...   \n",
       "9   Casohypotensin acts as a bradykinin-potentiati...   \n",
       "10  Casohypotensin acts as a bradykinin-potentiati...   \n",
       "11  Casohypotensin acts as a bradykinin-potentiati...   \n",
       "12  Casohypotensin acts as a bradykinin-potentiati...   \n",
       "13  Has opioid antagonist activity (PubMed:1369293...   \n",
       "14  Has opioid antagonist activity (PubMed:1369293...   \n",
       "15  Has opioid antagonist activity (PubMed:1369293...   \n",
       "16                                     indeterminable   \n",
       "17                                     indeterminable   \n",
       "18  Lactoferricin B is an antimicrobial peptide. I...   \n",
       "19  Lactoferricin B is an antimicrobial peptide. I...   \n",
       "20  Antioxidant peptide has 2,2-diphenyl-1-picrylh...   \n",
       "21      Antioxidant peptide has antioxidant activity.   \n",
       "\n",
       "                              non_associated_function  \\\n",
       "0                                                 NaN   \n",
       "1                                                 NaN   \n",
       "2                                                 NaN   \n",
       "3                                                 NaN   \n",
       "4                                                 NaN   \n",
       "5                                                 NaN   \n",
       "6                                                 NaN   \n",
       "7                                                 NaN   \n",
       "8                                                 NaN   \n",
       "9                                                 NaN   \n",
       "10                                                NaN   \n",
       "11                                                NaN   \n",
       "12                                                NaN   \n",
       "13                                                NaN   \n",
       "14                                                NaN   \n",
       "15                                                NaN   \n",
       "16  ['Kappa-casein stabilizes micelle formation, p...   \n",
       "17  ['Kappa-casein stabilizes micelle formation, p...   \n",
       "18                                                NaN   \n",
       "19                                                NaN   \n",
       "20                                                NaN   \n",
       "21                                                NaN   \n",
       "\n",
       "                                                title                authors  \\\n",
       "0   Potassium efflux induced by a new lactoferrin-...  Viejo-Diaz, M. et al.   \n",
       "1   Potassium efflux induced by a new lactoferrin-...  Viejo-Diaz, M. et al.   \n",
       "2   Effects of 'casoparan', a peptide isolated fro...      Lebrun, I. et al.   \n",
       "3   Effects of 'casoparan', a peptide isolated fro...      Lebrun, I. et al.   \n",
       "4   Effects of 'casoparan', a peptide isolated fro...      Lebrun, I. et al.   \n",
       "5   Casocidin-I: a casein-alpha s2 derived peptide...       Zucht, H. et al.   \n",
       "6   Biochemical and pharmacological aspects of two...    Perpetuo, E. et al.   \n",
       "7   Biochemical and pharmacological aspects of two...    Perpetuo, E. et al.   \n",
       "8   Biochemical and pharmacological aspects of two...    Perpetuo, E. et al.   \n",
       "9   Biochemical and pharmacological aspects of two...    Perpetuo, E. et al.   \n",
       "10  Isolation and characterization of a new bradyk...      Lebrun, I. et al.   \n",
       "11  Isolation and characterization of a new bradyk...      Lebrun, I. et al.   \n",
       "12  Isolation and characterization of a new bradyk...      Lebrun, I. et al.   \n",
       "13  Isolation and characterization of opioid antag...        Tani, F. et al.   \n",
       "14  Isolation and characterization of opioid antag...        Tani, F. et al.   \n",
       "15  Isolation and characterization of opioid antag...        Tani, F. et al.   \n",
       "16  Specificity of hydrolysis of bovine kappa-case...        Reid, J. et al.   \n",
       "17  Specificity of hydrolysis of bovine kappa-case...        Reid, J. et al.   \n",
       "18  Identification of the bactericidal domain of l...     Bellamy, W. et al.   \n",
       "19  Identification of the bactericidal domain of l...     Bellamy, W. et al.   \n",
       "20  Antioxidant peptides isolated from cheddar che...       Gupta, A. et al.   \n",
       "21  Studies on antioxidative peptides generated in...       Gupta, A. et al.   \n",
       "\n",
       "                                   doi   pubmed_id  \\\n",
       "0              10.1023/a:1022657630698  12693969.0   \n",
       "1              10.1023/a:1022657630698  12693969.0   \n",
       "2            10.1080/09629350400003068  15545057.0   \n",
       "3            10.1080/09629350400003068  15545057.0   \n",
       "4            10.1080/09629350400003068  15545057.0   \n",
       "5         10.1016/0014-5793(95)00974-e   7556666.0   \n",
       "6   10.1023/b:jopc.0000008724.98339.ff  14714726.0   \n",
       "7   10.1023/b:jopc.0000008724.98339.ff  14714726.0   \n",
       "8   10.1023/b:jopc.0000008724.98339.ff  14714726.0   \n",
       "9   10.1023/b:jopc.0000008724.98339.ff  14714726.0   \n",
       "10                     10.1139/y95-012   7600458.0   \n",
       "11                     10.1139/y95-012   7600458.0   \n",
       "12                     10.1139/y95-012   7600458.0   \n",
       "13             10.1271/bbb1961.54.1803   1369293.0   \n",
       "14             10.1271/bbb1961.54.1803   1369293.0   \n",
       "15             10.1271/bbb1961.54.1803   1369293.0   \n",
       "16       10.1128/aem.60.3.801-806.1994   8161175.0   \n",
       "17       10.1128/aem.60.3.801-806.1994   8161175.0   \n",
       "18        10.1016/0167-4838(92)90346-f   1599934.0   \n",
       "19        10.1016/0167-4838(92)90346-f   1599934.0   \n",
       "20                                 NaN         NaN   \n",
       "21                                 NaN         NaN   \n",
       "\n",
       "                                             abstract  \\\n",
       "0   A 31-amino acid synthetic peptide (NH(2)-FFSAS...   \n",
       "1   A 31-amino acid synthetic peptide (NH(2)-FFSAS...   \n",
       "2   Casein, a protein found in milk of several spe...   \n",
       "3   Casein, a protein found in milk of several spe...   \n",
       "4   Casein, a protein found in milk of several spe...   \n",
       "5   Here we report the isolation and characterizat...   \n",
       "6   Peptides that display bradykinin-potentiating ...   \n",
       "7   Peptides that display bradykinin-potentiating ...   \n",
       "8   Peptides that display bradykinin-potentiating ...   \n",
       "9   Peptides that display bradykinin-potentiating ...   \n",
       "10  Peptides that display bradykinin-potentiating ...   \n",
       "11  Peptides that display bradykinin-potentiating ...   \n",
       "12  Peptides that display bradykinin-potentiating ...   \n",
       "13  Peptides with affinity for opioid receptors we...   \n",
       "14  Peptides with affinity for opioid receptors we...   \n",
       "15  Peptides with affinity for opioid receptors we...   \n",
       "16  The cell envelope-associated proteinases from ...   \n",
       "17  The cell envelope-associated proteinases from ...   \n",
       "18  We report the existence of a previously unknow...   \n",
       "19  We report the existence of a previously unknow...   \n",
       "20                                                NaN   \n",
       "21                                                NaN   \n",
       "\n",
       "                           function  \\\n",
       "0                  Cell Penetrating   \n",
       "1                     Antimicrobial   \n",
       "2         'Bradykinin-Potentiating'   \n",
       "3                'Immunomodulatory'   \n",
       "4                  'Cytomodulatory'   \n",
       "5                     Antimicrobial   \n",
       "6                    ACE-inhibitory   \n",
       "7           Bradykinin-Potentiating   \n",
       "8                            Opioid   \n",
       "9                  Antihypertensive   \n",
       "10                 Antihypertensive   \n",
       "11          Bradykinin-Potentiating   \n",
       "12  Prolyl endopeptidase-inhibitory   \n",
       "13                           Opioid   \n",
       "14                           Opioid   \n",
       "15                           Opioid   \n",
       "16                 Antihypertensive   \n",
       "17                           Opioid   \n",
       "18                    Antimicrobial   \n",
       "19                        Cytotoxic   \n",
       "20                      Antioxidant   \n",
       "21                      Antioxidant   \n",
       "\n",
       "                                                logic  \\\n",
       "0   The ability of the peptide to induce the efflu...   \n",
       "1   The antimicrobial property is evident from the...   \n",
       "2   The peptide's ability to enhance the actions o...   \n",
       "3   The peptide's ability to increase the activity...   \n",
       "4   The peptide's direct influence on the activity...   \n",
       "5   The evidence for the antimicrobial function co...   \n",
       "6   This means that the peptide is competitive, wh...   \n",
       "7   The fact that the peptide potentiates the acti...   \n",
       "8   The observed effect implies that the peptide f...   \n",
       "9   This function results from the capacity of bra...   \n",
       "10  Hypotension, or reduced blood pressure, is ind...   \n",
       "11  The peptide's interaction with bradykinin, an ...   \n",
       "12  The peptide's ability to inhibit endo-oligopep...   \n",
       "13  The term opioid refers to compounds with narco...   \n",
       "14  Based on the evidence provided, this peptide d...   \n",
       "15  The peptide is directly described as having op...   \n",
       "16  As SRYPSY comes from Kappa-casein and given th...   \n",
       "17  As SRYPSY is a casoxin-6 peptide, it might pos...   \n",
       "18  The peptide has been proven to inhibit the gro...   \n",
       "19  The peptide causes a rapid loss of colony-form...   \n",
       "20  Given that the associated function is stated t...   \n",
       "21  The peptide is explicitly described as an anti...   \n",
       "\n",
       "                                             evidence  \n",
       "0   The antimicrobial effect of kaliocin-1 was low...  \n",
       "1   The peptide, termed kaliocin-1, and lactoferri...  \n",
       "2   \"It also acts as a bradykinin-potentiating pep...  \n",
       "3   \"Casoparan acts as a macrophage activator, inc...  \n",
       "4   \"The immunostimulating action is probably rela...  \n",
       "5   The abstract reports the isolation and charact...  \n",
       "6   The octapeptide YPVEPFTE showed an in vitro co...  \n",
       "7   The peptide YPVEPFTE has been described as a b...  \n",
       "8   The same peptide presented opiate-like activit...  \n",
       "9   Casohypotensin acts as a bradykinin-potentiati...  \n",
       "10  Intravenous injections of bradykinin and of th...  \n",
       "11  The octapeptide Tyr-Pro-Val-Gln-Pro-Phe-Thr-Gl...  \n",
       "12  This bradykinin-potentiating octapeptide is a ...  \n",
       "13  The peptide YLGSGY, also known as Lactoferroxi...  \n",
       "14  The peptide RYYGY (Lactoferroxin-B) has affini...  \n",
       "15  The peptide has been found to have an affinity...  \n",
       "16  The action of the cell envelope-associated pro...  \n",
       "17  Casoxins A, B, and C have opioid antagonist ac...  \n",
       "18  The description provided states \"Lactoferricin...  \n",
       "19  \"The effect of the isolated domain was lethal ...  \n",
       "20  The peptide HIQKEDVPSER is described as an ant...  \n",
       "21  The given peptide VKEAMAPK is described as an ...  "
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "updated_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "24921eca",
   "metadata": {},
   "outputs": [],
   "source": [
    "updated_df.to_csv('data/uni_reff_merged_final_me_gpt_df.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c659b2a4",
   "metadata": {},
   "source": [
    "# Compared Uniport list to MBPDB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad53406d",
   "metadata": {},
   "source": [
    "generate a list of unique peptides that can be cross referenced with MBPDB website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "e54b0c2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DVECGEGHFCHDNQTCCRDNRQGWACCPYRQGVCCADRRHCCPAGFRCAARGTKCL',\n",
       " 'LRQFLQKSLAAAA',\n",
       " 'RSSCFGGRMDRIGAQSGLGCNSFRY',\n",
       " 'SSCFGGRMDRIGAQSGLGCNSFR',\n",
       " 'LLGDFFRKSKEKIGKEFKRIVQRIKDFLR',\n",
       " 'QGPWLEEEEEAYGWMDF',\n",
       " 'SLRRSSCFGGRMDRIGAQSGLGCNSFRY',\n",
       " 'APRSLRRSSCFGGRMDRIGAQSGLG',\n",
       " 'AGCKNFFWKTFTSC',\n",
       " 'RSLQDTEEKSRSFSASQADPLSDPDQMNEDKRHSQGTFTSDYSKYLDSRRAQDFVQWLMNTKRNRNNIA',\n",
       " 'KYLGPQY',\n",
       " 'APLEPVYPGDNATPEQMAQYAADLRRYINMLTRPRY',\n",
       " 'FALLGDFFRKSKEKIGKEFKRIVQRIKDFLRNLVPRTES',\n",
       " 'HSQGTFTSDYSKYLDSRRAQDFVQWLMNTKRNRNNIA',\n",
       " 'YLGSGY',\n",
       " 'YGGFMRGL',\n",
       " 'SSCFGGRMDRIGAQSGLGCNSFRY',\n",
       " 'NPMYNAVSNADLMDFKNLLDHLEEKMPLED',\n",
       " 'HAEGTFTSDVSSYLEGQAAKEFIAWLVKGR',\n",
       " 'YPIKPEAPREDASPEELNRYYASLRHYLNLVTRQRY',\n",
       " 'HADGSFSDEMNTILDNLAARDFINWLIQTKITD',\n",
       " 'HSQGTFTSDYSKYLDSRRAQDFVQWLMNT',\n",
       " 'HAEGTFTSDVSSYLEGQAAKEFIAWLVKGRG',\n",
       " 'SSKITHRIHWESASLLR',\n",
       " 'GSSFLSPEHQRVQQRKESKKPPAKLQP',\n",
       " 'EVVPPQVLSEPNEEAGAALSPLPEVPPWTGEVSPAQR',\n",
       " 'YGGFM',\n",
       " 'YAEGTFISDYSIAMDKIHQQDFVNWLLAQKGKKNDWKHNITQ',\n",
       " 'RYYGY',\n",
       " 'AMVSEFLKQAWFIENEEQEYVQTVK',\n",
       " 'LTAPRSLRRSSCFGGRMDRIGAQSGLGCNSFRY',\n",
       " 'DHYNCVSSGGQCLYSACPIFTKIQGTCYRGKAKCCK',\n",
       " 'SSDRSALLKSKLRALLTAPR',\n",
       " 'YADAIFTNSYRKVLGQLSARKLLQDIMSRQQGESNQERGARARL',\n",
       " 'ACYCRIPACIAGERRYGTCIYQGRLWAFCC',\n",
       " 'RKSKEKIGKEFKRIVQRIKDFLRNLVPRTES',\n",
       " 'KCNTATCATQRLANFLVHSSNNFGAILSSTNVGSNTY',\n",
       " 'LLGDFFRKSKEKIGKEFKRIVQR',\n",
       " 'QLYENKPRRPYIL',\n",
       " 'VMCPDARSRCPDGSTCCELPSGKYGCCPMPNATCCSDHLHCCPQDTVCDLIQSKCL',\n",
       " 'LLGDFFRKSKEKIGKEFKRIVQRIKDFLRNLVPRTES',\n",
       " 'SSCFGGRMDRIGAQSGLGCNS',\n",
       " 'CYIQNCPLG',\n",
       " 'QGVNDNEEGFFSAR',\n",
       " 'HKEDTLAFSEWGSPHAAVPR',\n",
       " 'YGGFL',\n",
       " 'FFSASCVPGADKGQFPNLCRLCAGTGENKCA',\n",
       " 'HADGVFTSDFSKLLGQLSAKKYLESLMGKRVSSNISEDPVPV',\n",
       " 'ADSGEGDFLAEGGGVR',\n",
       " 'DQESCKGRCTEGFNVDKKCQCDELCSYYQSCCTDYTAECKPQVT',\n",
       " 'TAPRSLRRSSCFGGRMDRIGAQSGLGCNSFRY',\n",
       " 'GSSFLSPEHQRVQQRKESKKPPAKLQPR',\n",
       " 'KSKEKIGKEFKRIVQRIKDFLRNLVPRTES',\n",
       " 'CSCSSLMDKECVYFCHLDIIWVNTPEHVVPYGLGSPRS',\n",
       " 'HDEFERHAEGTFTSDVSSYLEGQAAKEFIAWLVKGRG',\n",
       " 'FFRKSKEKIGKEFKRIVQRIKDFLRNLVPRTES',\n",
       " 'RSSCFGGRMDRIGAQSGLGCNSFR',\n",
       " 'CSCSSLMDKECVYFCHLDIIW',\n",
       " 'KRIVQRIKDFLRNLVPRTES']"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set(reff_merged_final_me_gpt_df['peptide']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "169349e7",
   "metadata": {},
   "source": [
    "Creates new df with peptides discoved in Uniport search and match with MBPDB on peptide and DOI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "3ae60f3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of records from UniPort/GPT search: 697\n",
      "Number of peptides matched with MBPDB: 5\n",
      "Number of unique peptides not matched with MBPDB: 55\n",
      "Number of unique DOIs not matched with MBPDB: 85\n",
      "Number of unique functions from GPT not matched with MBPDB: 42\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteinID</th>\n",
       "      <th>chain</th>\n",
       "      <th>interval</th>\n",
       "      <th>peptide</th>\n",
       "      <th>description</th>\n",
       "      <th>associated_function</th>\n",
       "      <th>non_associated_function</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>doi</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>abstract</th>\n",
       "      <th>gpt_function</th>\n",
       "      <th>gpt_logic</th>\n",
       "      <th>gpt_evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>P02788</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>171-201</td>\n",
       "      <td>FFSASCVPGADKGQFPNLCRLCAGTGENKCA</td>\n",
       "      <td>Kaliocin-1</td>\n",
       "      <td>Has antimicrobial activity and is able to perm...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Potassium efflux induced by a new lactoferrin-...</td>\n",
       "      <td>Viejo-Diaz, M. et al.</td>\n",
       "      <td>10.1023/a:1022657630698</td>\n",
       "      <td>12693969</td>\n",
       "      <td>A 31-amino acid synthetic peptide (NH(2)-FFSAS...</td>\n",
       "      <td>Antimicrobial</td>\n",
       "      <td>The text evidence clearly mentions that the pe...</td>\n",
       "      <td>\"The peptide, termed kaliocin-1, and lactoferr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>P02788</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>171-201</td>\n",
       "      <td>FFSASCVPGADKGQFPNLCRLCAGTGENKCA</td>\n",
       "      <td>Kaliocin-1</td>\n",
       "      <td>Has antimicrobial activity and is able to perm...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Potassium efflux induced by a new lactoferrin-...</td>\n",
       "      <td>Viejo-Diaz, M. et al.</td>\n",
       "      <td>10.1023/a:1022657630698</td>\n",
       "      <td>12693969</td>\n",
       "      <td>A 31-amino acid synthetic peptide (NH(2)-FFSAS...</td>\n",
       "      <td>Cell Penetrating</td>\n",
       "      <td>The evidence suggests that the peptide can pas...</td>\n",
       "      <td>\"In addition, the peptide was able to mimic na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>P02788</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>338-343</td>\n",
       "      <td>YLGSGY</td>\n",
       "      <td>Lactoferroxin-A</td>\n",
       "      <td>Has opioid antagonist activity (PubMed:1369293...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolation and characterization of opioid antag...</td>\n",
       "      <td>Tani, F. et al.</td>\n",
       "      <td>10.1271/bbb1961.54.1803</td>\n",
       "      <td>1369293</td>\n",
       "      <td>Peptides with affinity for opioid receptors we...</td>\n",
       "      <td>Opioid</td>\n",
       "      <td>The term \"opioid antagonist activity\" directly...</td>\n",
       "      <td>According to the provided article, the peptide...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>P02788</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>543-547</td>\n",
       "      <td>RYYGY</td>\n",
       "      <td>Lactoferroxin-B</td>\n",
       "      <td>Has opioid antagonist activity (PubMed:1369293...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolation and characterization of opioid antag...</td>\n",
       "      <td>Tani, F. et al.</td>\n",
       "      <td>10.1271/bbb1961.54.1803</td>\n",
       "      <td>1369293</td>\n",
       "      <td>Peptides with affinity for opioid receptors we...</td>\n",
       "      <td>Opioid</td>\n",
       "      <td>The peptide has been shown to interact with op...</td>\n",
       "      <td>The peptide in question, RYYGY, which correspo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>P02788</td>\n",
       "      <td>Lactotransferrin</td>\n",
       "      <td>680-686</td>\n",
       "      <td>KYLGPQY</td>\n",
       "      <td>Lactoferroxin-C</td>\n",
       "      <td>Has opioid antagonist activity (PubMed:1369293...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Isolation and characterization of opioid antag...</td>\n",
       "      <td>Tani, F. et al.</td>\n",
       "      <td>10.1271/bbb1961.54.1803</td>\n",
       "      <td>1369293</td>\n",
       "      <td>Peptides with affinity for opioid receptors we...</td>\n",
       "      <td>Opioid</td>\n",
       "      <td>The bioactivity functions were derived from th...</td>\n",
       "      <td>Peptides derived from human lactoferrin, inclu...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    proteinID             chain interval                          peptide  \\\n",
       "135    P02788  Lactotransferrin  171-201  FFSASCVPGADKGQFPNLCRLCAGTGENKCA   \n",
       "135    P02788  Lactotransferrin  171-201  FFSASCVPGADKGQFPNLCRLCAGTGENKCA   \n",
       "136    P02788  Lactotransferrin  338-343                           YLGSGY   \n",
       "137    P02788  Lactotransferrin  543-547                            RYYGY   \n",
       "138    P02788  Lactotransferrin  680-686                          KYLGPQY   \n",
       "\n",
       "         description                                associated_function  \\\n",
       "135       Kaliocin-1  Has antimicrobial activity and is able to perm...   \n",
       "135       Kaliocin-1  Has antimicrobial activity and is able to perm...   \n",
       "136  Lactoferroxin-A  Has opioid antagonist activity (PubMed:1369293...   \n",
       "137  Lactoferroxin-B  Has opioid antagonist activity (PubMed:1369293...   \n",
       "138  Lactoferroxin-C  Has opioid antagonist activity (PubMed:1369293...   \n",
       "\n",
       "    non_associated_function  \\\n",
       "135                     NaN   \n",
       "135                     NaN   \n",
       "136                     NaN   \n",
       "137                     NaN   \n",
       "138                     NaN   \n",
       "\n",
       "                                                 title                authors  \\\n",
       "135  Potassium efflux induced by a new lactoferrin-...  Viejo-Diaz, M. et al.   \n",
       "135  Potassium efflux induced by a new lactoferrin-...  Viejo-Diaz, M. et al.   \n",
       "136  Isolation and characterization of opioid antag...        Tani, F. et al.   \n",
       "137  Isolation and characterization of opioid antag...        Tani, F. et al.   \n",
       "138  Isolation and characterization of opioid antag...        Tani, F. et al.   \n",
       "\n",
       "                         doi pubmed_id  \\\n",
       "135  10.1023/a:1022657630698  12693969   \n",
       "135  10.1023/a:1022657630698  12693969   \n",
       "136  10.1271/bbb1961.54.1803   1369293   \n",
       "137  10.1271/bbb1961.54.1803   1369293   \n",
       "138  10.1271/bbb1961.54.1803   1369293   \n",
       "\n",
       "                                              abstract      gpt_function  \\\n",
       "135  A 31-amino acid synthetic peptide (NH(2)-FFSAS...     Antimicrobial   \n",
       "135  A 31-amino acid synthetic peptide (NH(2)-FFSAS...  Cell Penetrating   \n",
       "136  Peptides with affinity for opioid receptors we...            Opioid   \n",
       "137  Peptides with affinity for opioid receptors we...            Opioid   \n",
       "138  Peptides with affinity for opioid receptors we...            Opioid   \n",
       "\n",
       "                                             gpt_logic  \\\n",
       "135  The text evidence clearly mentions that the pe...   \n",
       "135  The evidence suggests that the peptide can pas...   \n",
       "136  The term \"opioid antagonist activity\" directly...   \n",
       "137  The peptide has been shown to interact with op...   \n",
       "138  The bioactivity functions were derived from th...   \n",
       "\n",
       "                                          gpt_evidence  \n",
       "135  \"The peptide, termed kaliocin-1, and lactoferr...  \n",
       "135  \"In addition, the peptide was able to mimic na...  \n",
       "136  According to the provided article, the peptide...  \n",
       "137  The peptide in question, RYYGY, which correspo...  \n",
       "138  Peptides derived from human lactoferrin, inclu...  "
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Step 1: Create db_match_df\n",
    "\n",
    "# Filter rows from uni_reff_merged_final_df where the 'peptide' matches with those in mbpdb_df\n",
    "# We first convert the relevant columns of mbpdb_df to a dictionary with 'list' orientation.\n",
    "# Then, we use the 'isin' method to check if the values in the respective columns of uni_reff_merged_final_df are present in this dictionary.\n",
    "db_match_df = reff_merged_final_me_gpt_df[\n",
    "    reff_merged_final_me_gpt_df[['peptide']].isin(\n",
    "        mbpdb_df[['peptide']].to_dict(orient='list')\n",
    "    ).all(axis=1)\n",
    "]\n",
    "\n",
    "# Number of unique peptides, DOIs, and functions not matched with MBPDB\n",
    "unique_peptides_not_matched = reff_merged_final_me_gpt_df[~reff_merged_final_me_gpt_df['peptide'].isin(db_match_df['peptide'])]['peptide'].nunique()\n",
    "unique_doi_not_matched = reff_merged_final_me_gpt_df[~reff_merged_final_me_gpt_df['doi'].isin(db_match_df['doi'])]['doi'].nunique()\n",
    "unique_gpt_functions_not_matched = reff_merged_final_me_gpt_df[~reff_merged_final_me_gpt_df['gpt_function'].isin(db_match_df['gpt_function'])]['gpt_function'].nunique()\n",
    "\n",
    "# Print the number of rows in the original dataframe uni_reff_merged_final_df\n",
    "print(\"Number of records from UniPort/GPT search:\", reff_merged_final_me_gpt_df.shape[0])\n",
    "\n",
    "# Print the number of rows in the filtered dataframe db_match_df\n",
    "print(\"Number of peptides matched with MBPDB:\", db_match_df.shape[0])\n",
    "\n",
    "# Print the number of unique peptides, DOIs, and functions not matched with MBPDB\n",
    "print(\"Number of unique peptides not matched with MBPDB:\", unique_peptides_not_matched)\n",
    "print(\"Number of unique DOIs not matched with MBPDB:\", unique_doi_not_matched)\n",
    "print(\"Number of unique functions from GPT not matched with MBPDB:\", unique_gpt_functions_not_matched)\n",
    "\n",
    "# Display the resulting db_match_df\n",
    "db_match_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "78fb59d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Antithrombotic',\n",
       " 'Bacterial growth promoting',\n",
       " 'Immunomodulatory',\n",
       " 'Cell Penetrating',\n",
       " 'Increase exocrine pancreatic secretion',\n",
       " 'DPP-IV Inhibitory',\n",
       " 'Increase mucin secretion',\n",
       " 'DNA repair',\n",
       " 'Antimicrobial',\n",
       " 'Antioxidant',\n",
       " 'Nitric oxide liberation',\n",
       " 'Ameliorates insulin resistance',\n",
       " 'Increase calcium uptake',\n",
       " 'Satiety',\n",
       " 'Opioid',\n",
       " 'Improves cognition',\n",
       " 'Cytomodulatory',\n",
       " 'Increase cellular growth',\n",
       " 'Antihypertensive',\n",
       " 'Prolyl endopeptidase-inhibitory',\n",
       " 'ACE-inhibitory',\n",
       " 'Increase intestinal motility',\n",
       " 'Cathepsin B Inhibitory',\n",
       " 'Cytotoxic',\n",
       " 'Bradykinin-Potentiating',\n",
       " 'Enhance insulin signaling',\n",
       " 'Wound healing',\n",
       " 'Antiapoptotic effect',\n",
       " 'Antianxiety',\n",
       " 'Osteoanabolic',\n",
       " 'Anticancer',\n",
       " 'Cholesterol regulation']"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mbpdb_function_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6398a12",
   "metadata": {},
   "source": [
    "Creates dataframe with peptides that did not match MBPDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7870b936",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique peptides w/o MBPDB match: 9\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b1009e52572c46af9fa9ecfce72b16f7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': False, 'defa…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Step 2: Create no_db_match_df\n",
    "\n",
    "# To identify rows in uni_reff_merged_final_df that are not in db_match_df, we drop rows in uni_reff_merged_final_df\n",
    "# that have the same index as the rows in db_match_df. The resulting dataframe, no_db_match_df, contains \n",
    "# rows from uni_reff_merged_final_df that didn't match with mbpdb_df based on 'peptide' and 'doi'.\n",
    "no_db_match_df = uni_reff_merged_final_df.drop(db_match_df.index)\n",
    "\n",
    "# Print the number of rows in the resulting dataframe no_db_match_df\n",
    "print(\"Number of unique peptides w/o MBPDB match:\", no_db_match_df.shape[0])\n",
    "\n",
    "# Display no_db_match_df using qgrid for an interactive view\n",
    "# The 'show_toolbar' option enables a toolbar in the qgrid display, and the 'forceFitColumns' option in grid_options \n",
    "# ensures that columns are not forcibly fit into the grid's width, allowing horizontal scrolling.\n",
    "no_db_match_df_qgrid = qgrid.show_grid(no_db_match_df, show_toolbar=True, grid_options={'forceFitColumns': False})\n",
    "\n",
    "# Return the qgrid display\n",
    "no_db_match_df_qgrid\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "490ec600",
   "metadata": {},
   "source": [
    "# Downloads a XML file from UniProt or PubMed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42b01db8",
   "metadata": {},
   "source": [
    "UniPort page download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b4a216ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the protein ID\n",
    "protein = 'P10451'\n",
    "\n",
    "# Construct the URL using the protein ID to fetch the XML content from UniProt\n",
    "url = f'https://www.uniprot.org/uniprot/{protein}.xml'\n",
    "\n",
    "# Use the 'requests' library to get the content of the URL\n",
    "response = requests.get(url)\n",
    "    \n",
    "# Check to ensure the request was successful (HTTP status code 200 means \"OK\")\n",
    "if response.status_code == 200:\n",
    "    # If successful, save the XML content to a local file with the name '{protein}.xml'\n",
    "    with open(f'xml_examples/{protein}.xml', 'wb') as file:\n",
    "        file.write(response.content)\n",
    "else:\n",
    "    # If the request wasn't successful, print an error message with the HTTP status code\n",
    "    print(f'Failed to retrieve the file: {response.status_code}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc56a30f",
   "metadata": {},
   "source": [
    "PubMed page download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "928c5d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the PubMed ID\n",
    "pubmed_id = '1369293'\n",
    "\n",
    "# Use the Entrez system (from the BioPython library) to fetch the XML details of the article with the specified PubMed ID\n",
    "handle = Entrez.efetch(db=\"pubmed\", id=pubmed_id, retmode=\"xml\")\n",
    "\n",
    "# Save the raw XML response to a local file named '{pubmed_id}.xml'\n",
    "with open(f'xml_examples/{pubmed_id}.xml', 'wb') as file:\n",
    "    file.write(handle.read())  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
